// mews-ui/src/Dialog/Dialog.jsx
import _ from 'lodash';
import React, { Component } from 'react';
import { PropTypes as pt } from 'prop-types';
import { DialogContainerElement, BackdropElement, DialogElement, DialogContentElement, DialogActionsElement, DialogButton } from './style';

class Dialog extends Component {
    static propTypes = {
        actions: pt.array,
        onBackdropClick: pt.func,
        opened: pt.bool,
    };

    render() {
        const { actions, children, opened, onBackdropClick, ...other } = this.props;

        const backdropElementProps = {
            isVisible: opened,
            onClick: onBackdropClick,
        };

        const dialogElementProps = {
            isOpened: opened,
        };

        return (
            <DialogContainerElement {...other}>
                {opened && <BackdropElement {...backdropElementProps} />}
                {opened && (
                    <DialogElement {...dialogElementProps}>
                        <DialogContentElement>
                            {children}
                        </DialogContentElement>
                        <DialogActionsElement>
                            {_.map(actions, (a, index) => <DialogButton key={index} {...a} />)}
                        </DialogActionsElement>
                    </DialogElement>
                )}
            </DialogContainerElement>
        );
    }
}

export default Dialog;
