// mews-ui/src/Dialog/style.jsx
import styled from 'styled-components';
import { PALETTE_GREY_900, COLOR_WHITE } from '../colors';
import { ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, Z_DEPTH_SHADOW_3, Z_INDEX_HIGH } from '../variables';
import StyledTransitionGroup from '../TransitionGroup/StyledTransitionGroup';
import Button from '../Button/Button';

export const DialogContainerElement = styled(StyledTransitionGroup)`
    position: fixed;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    overflow: hidden;

    display: flex;
    align-items: center;
    justify-content: center;

    pointer-events: none;
    z-index: ${Z_INDEX_HIGH};
`;

export const BackdropElement = styled.div`
    position: absolute;
    width: 100%;
    height: 100%;
    background-color: ${PALETTE_GREY_900};
    transition: opacity ${ANIMATION_DURATION} ${ANIMATION_CURVE_DEFAULT};

    pointer-events: initial;

    ${props => props.enter && `opacity: 0;`}
    ${props => props.enterActive && `opacity: .48;`}

    ${props => props.leave && `opacity: .48;`}
    ${props => props.leaveActive && `opacity: 0;`}
`;

export const DialogElement = styled.div`
    background-color: ${COLOR_WHITE};
    box-shadow: ${Z_DEPTH_SHADOW_3};
    z-index: 1;
    width: 50vw;
    max-width: 98vw;
    max-height: 98vh;

    pointer-events: initial;

    transition-duration: ${ANIMATION_DURATION};
    transition-timing-function: ${ANIMATION_CURVE_DEFAULT};
    transition-property: opacity, transform;

    ${props => props.enter && `
        opacity: 0;
        transform: translateY(-4em);
    `}
    ${props => props.enterActive && `
        opacity: 1;
        transform: translateY(0);
   `}

    ${props => props.leave && `
        opacity: 1;
        transform: translateY(0);
    `}
    ${props => props.leaveActive && `
        opacity: 0;
        transform: translateY(-4em);
    `}
`;

export const DialogContentElement = styled.div`
    padding: 24px;
`;

export const DialogActionsElement = styled.div`
    padding: 8px;
    text-align: right;
`;

export const DialogButton = styled(Button)`
    &:not(:last-child) {
        margin-right: 8px;
    }
`;
