// mews-ui/lib/utils/events.js
import _ from 'lodash';

export function getMousePosition(event) {
    return {
        x: event.pageX - (window.scrollX || window.pageXOffset),
        y: event.pageY - (window.scrollY || window.pageYOffset)
    };
}

export function getTouchPosition(event) {
    return {
        x: event.touches[0].pageX - (window.scrollX || window.pageXOffset),
        y: event.touches[0].pageY - (window.scrollY || window.pageYOffset)
    };
}

export function addEventsToDocument(eventMap) {
    _.each(eventMap, function (handler, type) {
        return document.addEventListener(type, handler, false);
    });
}

export function removeEventsFromDocument(eventMap) {
    _.each(eventMap, function (handler, type) {
        return document.removeEventListener(type, handler, false);
    });
}

export function addEventListenerOnTransitionEnded(element, fn) {
    if (!element) {
        return false;
    }

    element.addEventListener('transitionend', fn);
    return true;
}

export function removeEventListenerOnTransitionEnded(element, fn) {
    if (!element) {
        return false;
    }

    element.removeEventListener('transitionend', fn);
    return true;
}

export function isTargetDescendant(event, parent) {
    return isDescendant(event.target, parent);
}

export function isDescendant(element, parent) {
    var node = element;
    while (node !== null) {
        if (node === parent) {
            return true;
        }
        node = node.parentNode;
    }

    return false;
}

export function pauseEvent(event) {
    event.stopPropagation();
    event.preventDefault();
}
