// mews-ui/src/Switch/style.jsx
import styled from 'styled-components';
import { COLOR_BLACK, COLOR_WHITE } from '../colors';
import { ANIMATION_CURVE_DEFAULT, DEFAULT_THEME, SHADOW_2P } from '../variables';

const SWITCH_WIDTH = 34;
const SWITCH_HEIGHT = 14;
const KNOB_SIZE = 20;

export const SwitchElement = styled.div`
    display: inline-block;
    position: relative;
`;

export const NativeControlElement = styled.input.attrs({
    type: 'checkbox',
})`
    width: ${SWITCH_WIDTH}px;
    height: ${SWITCH_HEIGHT}px;
    cursor: pointer;
    position: absolute;
    padding: 0;
    margin: 0;
    top: 0;
    left: 0;
    box-sizing: border-box;
    opacity: 0;
    z-index: 2;
`;

export const SwitchBackgroundElement = styled.div`
    cursor: pointer;
    border-radius: ${SWITCH_HEIGHT / 2}px;
    background-color: transparent;
    height: ${SWITCH_HEIGHT}px;
    outline: none;
    position: relative;
    width: ${SWITCH_WIDTH}px;
    user-select: none;

    &:before {
        display: block;
        position: absolute;
        left: 0;
        right: 0;
        top: 0;
        bottom: 0;
        border-radius: ${SWITCH_HEIGHT / 2}px;
        content: '';
        background-color: ${COLOR_BLACK};
        opacity: .38;

        transition-duration: 90ms;
        transition-property: opacity, background-color;
        transition-timing-function: ${ANIMATION_CURVE_DEFAULT};

        ${props => props.checked && `
            background-color: ${props.theme.primaryColor};
            opacity: .5;
        `}
    }
`;
SwitchBackgroundElement.defaultProps = { theme: DEFAULT_THEME };

export const SwitchKnobElement = styled.div`
    border-radius: 50%;
    height: ${KNOB_SIZE}px;
    width: ${KNOB_SIZE}px;
    box-shadow: ${SHADOW_2P};

    position: absolute;
    top: ${(SWITCH_HEIGHT - KNOB_SIZE) / 2}px;

    transition-duration: 90ms;
    transition-property: transform, background-color;
    transition-timing-function: ${ANIMATION_CURVE_DEFAULT};

    background-color: ${COLOR_WHITE};
    transform: translateX(0);

    z-index: 1;

    ${props => props.checked && `
        background-color: ${props.theme.primaryColor};
        transform: translateX(${SWITCH_WIDTH - KNOB_SIZE}px);
    `}

`;
SwitchKnobElement.defaultProps = { theme: DEFAULT_THEME };
