// mews-ui/src/Switch/Switch.jsx
import _ from 'lodash';
import React, { PropTypes as pt } from 'react';
import { noop } from '../utils/func';
import { SwitchElement, NativeControlElement, SwitchBackgroundElement, SwitchKnobElement } from './style';

class Switch extends React.Component {
    static propTypes = {
        checked: pt.bool,
        disabled: pt.bool,
        onChange: pt.func,
    };

    static defaultProps = {
        checked: false,
        disabled: false,
        onChange: noop,
    };

    state = {
        checked: this.props.checked,
    };

    componentWillReceiveProps(nextProps) {
        if (!_.isUndefined(nextProps.checked) && nextProps.checked !== this.props.checked) {
            this.setState({ checked: nextProps.checked });
        }
    }

    handleToggle = () => {
        if (this.props.disabled) {
            return;
        }

        this.toggleCheck();
    };

    toggleCheck = () => {
        this.setState({ checked: !this.state.checked }, () => this.props.onChange(this.state.checked));
    };

    render() {
        const { disabled } = this.props;
        const { checked } = this.state;

        const nativeControlProps = {
            checked,
            onClick: this.handleToggle,
            readOnly: true,
        };

        const switchProps = { disabled, checked };

        return (
            <SwitchElement>
                <NativeControlElement {...nativeControlProps} />
                <SwitchBackgroundElement {...switchProps}>
                    <SwitchKnobElement {...switchProps} />
                </SwitchBackgroundElement>
            </SwitchElement>
        );
    }
}

export default Switch;
