// mews-ui/lib/variables.js
import { COLOR_DARK_CONTRAST, COLOR_BLACK, PALETTE_BLUE_500, PALETTE_GREY_200, PALETTE_GREY_600, PALETTE_GREY_900 } from './colors';
import { alpha } from './utils/color';

export var COLOR_DIVIDER = PALETTE_GREY_200;
export var COLOR_TEXT = PALETTE_GREY_900;
export var COLOR_TEXT_SECONDARY = PALETTE_GREY_600;

export var COLOR_PRIMARY = PALETTE_BLUE_500;
export var COLOR_PRIMARY_CONTRAST = COLOR_DARK_CONTRAST;

export var PREFERRED_FONT = "'Roboto', 'Helvetica', 'Arial', sans-serif";
export var FONT_SIZE_TINY = '12px';
export var FONT_SIZE_SMALL = '14px';
export var FONT_SIZE_NORMAL = '16px';
export var FONT_SIZE_LARGE = '25px';
export var FONT_SIZE = FONT_SIZE_NORMAL;
export var FONT_WEIGHT_THIN = 300;

export var ANIMATION_DURATION = '0.35s';
export var ANIMATION_CURVE_FAST_OUT_LINEAR_IN = 'cubic-bezier(.4, 0, 1, 1)';
export var ANIMATION_CURVE_FAST_OUT_SLOW_IN = 'cubic-bezier(.4, 0, .2, 1)';
export var ANIMATION_CURVE_LINEAR_OUT_SLOW_IN = 'cubic-bezier(0, 0, .2, 1)';
export var ANIMATION_CURVE_DEFAULT = ANIMATION_CURVE_FAST_OUT_SLOW_IN;

export var SHADOW_2P = '\n    0 2px 2px 0 rgba(0, 0, 0, 0.14),\n    0 3px 1px -2px rgba(0, 0, 0, 0.2),\n    0 1px 5px 0 rgba(0, 0, 0, 0.12);\n';

export var SHADOW_3P = '\n    0 3px 4px 0 rgba(0, 0, 0, 0.14),\n    0 3px 3px -2px rgba(0, 0, 0, 0.2),\n    0 1px 8px 0 rgba(0, 0, 0, 0.12);\n';

export var Z_DEPTH_SHADOW_1 = '0 1px 6px rgba(0, 0, 0, 0.12), 0 1px 4px rgba(0, 0, 0, 0.24)';
export var Z_DEPTH_SHADOW_2 = '0 3px 10px rgba(0, 0, 0, 0.16), 0 3px 10px rgba(0, 0, 0, 0.23)';
export var Z_DEPTH_SHADOW_3 = '0 10px 30px rgba(0, 0, 0, 0.19), 0 6px 10px rgba(0, 0, 0, 0.23)';
export var Z_DEPTH_SHADOW_4 = '0 14px 45px rgba(0, 0, 0, 0.25), 0 10px 18px rgba(0, 0, 0, 0.22)';
export var Z_DEPTH_SHADOW_5 = '0 19px 60px rgba(0, 0, 0, 0.3), 0 15px 20px rgba(0, 0, 0, 0.22)';

export var Z_INDEX_HIGHEST = 300;
export var Z_INDEX_HIGHER = 200;
export var Z_INDEX_HIGH = 100;
export var Z_INDEX_NORMAL = 1;
export var Z_INDEX_LOW = -100;
export var Z_INDEX_LOWER = -200;

var disabledColor = alpha(COLOR_BLACK, 0.38);
var disabledSwitchColor = '#fafafa';

export var DEFAULT_THEME = {
    primaryColor: COLOR_PRIMARY,
    primaryContrastColor: COLOR_PRIMARY_CONTRAST,

    primaryTextColor: alpha(COLOR_BLACK, 0.87),
    secondaryTextColor: alpha(COLOR_BLACK, 0.54),
    disabledColor: disabledColor,
    disabledSwitchColor: disabledSwitchColor,
    hintTextColor: disabledColor,
    iconColor: disabledColor,
    dividerColor: alpha(COLOR_BLACK, 0.12)
};
