// mews-ui/src/Sidebar/style.jsx
import styled, { css } from 'styled-components';
import { PALETTE_GREY_900, COLOR_WHITE } from '../colors';
import { ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, Z_DEPTH_SHADOW_2, DEFAULT_THEME, Z_INDEX_HIGH } from '../variables';
import StyledTransitionGroup from '../TransitionGroup/StyledTransitionGroup';

export const SidebarContainerElement = styled(StyledTransitionGroup)`
    min-width: 320px;
    position: fixed;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    overflow: hidden;
    pointer-events: none;
    z-index: ${Z_INDEX_HIGH};
`;

export const BackdropElement = styled.div`
    position: absolute;
    width: 100%;
    height: 100%;
    background-color: ${PALETTE_GREY_900}; 
    transition: opacity ${ANIMATION_DURATION} ${ANIMATION_CURVE_DEFAULT};
    opacity: 0;

    ${props => props.enter && `opacity: 0;`}
    ${props => props.enterActive && `opacity: .48;`}

    ${props => props.leave && `opacity: .48;`}
    ${props => props.leaveActive && `opacity: 0;`}

    ${props => props.opened && css`
        pointer-events: initial;
    `};
`;

export const SidebarElement = styled.div`
    min-width: 320px;
    width: ${props => props.width};
    height: 100%;
    position: absolute;
    top: 0;
    right: 0;
    pointer-events: initial;
    display: flex;
    flex-direction: column;
    background-color: ${COLOR_WHITE};
    transition: transform ${ANIMATION_DURATION} ${ANIMATION_CURVE_DEFAULT};

    ${props => props.enter && `transform: translateX(100%);`}
    ${props => props.enterActive && `transform: translateX(0);`}

    ${props => props.leave && `transform: translateX(0);`}
    ${props => props.leaveActive && `transform: translateX(100%);`}

    ${props => props.opened && css`
        box-shadow: ${Z_DEPTH_SHADOW_2};
    `}
`;

export const SidebarHeaderElement = styled.div`
    display: flex;
    align-items: center;
    flex-shrink: 0;
    border-bottom: 1px solid;
    border-color: ${props => props.theme.dividerColor};
    box-sizing: border-box;
    padding-right: 16px;
`;
SidebarHeaderElement.defaultProps = { theme: DEFAULT_THEME };

export const SidebarHeaderTitleElement = styled.div`
    white-space: nowrap;
    text-overflow: ellipsis;
`;

export const SidebarContentElement = styled.div`
    flex-grow: 1;
    overflow-y: auto;
    overflow-x: hidden;
    padding: 16px;
`;
