// mews-js/lib/utils/semaphore.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = semaphore;

var _eventemitter = require('eventemitter3');

var _eventemitter2 = _interopRequireDefault(_eventemitter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var event = {
    TAKE: 'take',
    LEAVE: 'leave'
};

function semaphore() {
    var capacity = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;

    var emitter = new _eventemitter2.default();
    var pending = 0;

    return {
        take: function take() {
            if (pending < capacity) {
                emitter.emit(event.TAKE);
            }
            ++pending;
        },
        leave: function leave() {
            --pending;
            if (pending < capacity) {
                emitter.emit(event.LEAVE);
            }
        },
        onTake: function onTake(callback) {
            emitter.on(event.TAKE, callback);
        },
        onLeave: function onLeave(callback) {
            emitter.on(event.LEAVE, callback);
        }
    };
}
