// mews-ui/src/Calendar/CalendarMonth.jsx
import _ from 'lodash';
import React, { PureComponent } from 'react';
import { PropTypes as pt } from 'prop-types';

import CalendarDay from './CalendarDay';
import { MonthElement, MonthNameElement, WeekElement } from './style';

export default class Month extends PureComponent {
    static propTypes = {
        month: pt.any.isRequired, // eslint-disable-line react/forbid-prop-types
        getCalendarService: pt.func.isRequired,
    };

    static defaultProps = {
    };

    constructor(props) {
        super(props);
        const { month, getCalendarService } = props;

        this.state = {
            weeks: getCalendarService().getWeeksOfMonth(month),
        };
    }

    render() {
        const { month, getCalendarService, ...other } = this.props;
        const { weeks } = this.state;

        const monthName = getCalendarService().getMonthName(month);

        return (
            <MonthElement>
                <MonthNameElement>
                    {monthName}
                </MonthNameElement>
                {_.map(weeks, (week, i) => (
                    <WeekElement key={i}>
                        {_.map(week, (day, j) => (
                            <CalendarDay
                                key={j}
                                day={day}
                                getCalendarService={getCalendarService}
                                {...other}
                            />
                        ))}
                    </WeekElement>
                ))}
            </MonthElement>
        );
    }
}
