// mews-js/lib/components/finance/paymentGateways/braintreeComponent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = BraintreeComponent;

var _client = require('braintree-web/client');

var _client2 = _interopRequireDefault(_client);

var _utils = require('../../../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function BraintreeComponent() {
    return _utils.type.create(BraintreeComponent, {
        tokenizeCard: function tokenizeCard(_ref, _ref2) {
            var cvv = _ref.cvv,
                number = _ref.number,
                _ref$expiration = _ref.expiration,
                expirationMonth = _ref$expiration.month,
                expirationYear = _ref$expiration.year;
            var clientToken = _ref2.clientToken;

            return new Promise(function (resolve, reject) {
                _client2.default.create({ authorization: clientToken }, function (clientErr, clientInstance) {
                    if (clientErr) {
                        return reject(clientErr);
                    }

                    clientInstance.request({
                        endpoint: 'payment_methods/credit_cards',
                        method: 'post',
                        data: {
                            creditCard: {
                                cvv: cvv,
                                expirationMonth: expirationMonth,
                                expirationYear: expirationYear,
                                number: number
                            }
                        }
                    }, function (err, response) {
                        if (err) {
                            return reject(err);
                        }

                        resolve(response.creditCards[0].nonce);
                    });
                });
            });
        }
    });
}
