// mews-js/lib/components/finance/paymentGateways/stripeComponent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = StripeComponent;

var _utils = require('../../../utils');

var STRIPE_LIBRARY_URL = 'https://js.stripe.com/v2';
var DUMMY_CVC = '123';

function StripeComponent() {
    return _utils.type.create(StripeComponent, {
        tokenizeCard: function tokenizeCard(_ref, _ref2) {
            var cvc = _ref.cvv,
                number = _ref.number,
                _ref$expiration = _ref.expiration,
                expirationMonth = _ref$expiration.month,
                expirationYear = _ref$expiration.year;
            var publishableKey = _ref2.publishableKey;

            return Promise.resolve().then(function () {
                return global.Stripe || (0, _utils.loadScript)(STRIPE_LIBRARY_URL);
            }).then(function () {
                global.Stripe.setPublishableKey(publishableKey);

                return new Promise(function (resolve, reject) {
                    global.Stripe.card.createToken({
                        number: number,
                        cvc: cvc || DUMMY_CVC,
                        exp_month: expirationMonth, // eslint-disable-line camelcase
                        exp_year: expirationYear }, function (status, response) {
                        if (response.error) {
                            return reject(response.error);
                        }

                        resolve(response.id);
                    });
                });
            });
        }
    });
}
