// mews-ui/src/Tooltip/style.jsx
import styled from 'styled-components';
import { ANIMATION_CURVE_DEFAULT, PREFERRED_FONT, Z_INDEX_HIGHER } from '../variables';
import { COLOR_WHITE } from '../colors';
import Position from './Position';

export const TooltipElement = styled.div`
    display: block;
    font-family: ${PREFERRED_FONT};
    font-size: 12px;
    font-weight: 500;
    max-width: 200px;
    padding: 4px;
    pointer-events: none;
    position: absolute;
    text-align: center;
    text-transform: none;
    transform: scale(0) translateX(-50%);
    transform-origin: top left;
    transition: ${ANIMATION_CURVE_DEFAULT} .2s transform;
    z-index: ${Z_INDEX_HIGHER};

    left: ${props => props.left}px;
    top: ${props => props.top}px;

    ${props => props.active && `
        transform: scale(1) translateX(-50%);
    `}

    ${props => props.position === Position.Top && `
        transform: scale(0) translateX(-50%) translateY(-100%);

        ${props.active && 'transform: scale(1) translateX(-50%) translateY(-100%);'};
    `}

    ${props => props.position === Position.Left && `
        transform: scale(0) translateX(-100%) translateY(-50%);

        ${props.active && 'transform: scale(1) translateX(-100%) translateY(-50%);'};
      ;
    `}

    ${props => props.position === Position.Right && `
        transform: scale(0) translateX(0) translateY(-50%);

        ${props.active && 'transform: scale(1) translateX(0) translateY(-50%);'};
      ;
    `}
`;

export const TooltipContentElement = styled.div`
    background: rgba(0, 0, 0, .5);
    border-radius: 2px;
    color: ${COLOR_WHITE};
    padding: 8px;
`;

// arrow style
/*
    content: '';
    bottom: 0;
    left: 50%;
    margin-left: -5px;
    border-width: 5px 5px 0;
    border-color: transparent;
    border-top-color: #000;
*/
