// mews-ui/lib/Button/Button.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import { PropTypes as pt } from 'prop-types';

import { noop } from '../utils/func';
import { DEFAULT_THEME } from '../variables';
import { ButtonElement, ButtonHrefElement, ButtonIconElement } from './style';

var Button = function (_Component) {
    _inherits(Button, _Component);

    function Button() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Button);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Button.__proto__ || Object.getPrototypeOf(Button)).call.apply(_ref, [this].concat(args))), _this), _this.handleMouseUp = function (e) {
            _this.button.blur();
            _this.props.onMouseUp(e);
        }, _this.handleMouseLeave = function (e) {
            _this.button.blur();
            _this.props.onMouseLeave(e);
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Button, [{
        key: 'render',
        value: function render() {
            var _this2 = this;

            var _props = this.props,
                children = _props.children,
                flat = _props.flat,
                floating = _props.floating,
                href = _props.href,
                icon = _props.icon,
                _innerRef = _props.innerRef,
                label = _props.label,
                mini = _props.mini,
                primary = _props.primary,
                type = _props.type,
                raised = _props.raised,
                others = _objectWithoutProperties(_props, ['children', 'flat', 'floating', 'href', 'icon', 'innerRef', 'label', 'mini', 'primary', 'type', 'raised']);

            var Element = href ? ButtonHrefElement : ButtonElement;

            Element.defaultProps = {
                theme: DEFAULT_THEME
            };

            var level = primary ? 'primary' : 'neutral';

            var shape = 'flat';
            if (flat) {
                shape = 'flat';
            } else if (raised) {
                shape = 'raised';
            } else if (floating) {
                shape = 'floating';
            }

            var props = _extends({}, others, {
                disabled: this.props.disabled,
                icon: !!icon,
                innerRef: function innerRef(c) {
                    _this2.button = c;
                    _innerRef(c);
                },
                href: href,
                mini: mini,
                onMouseLeave: this.handleMouseLeave,
                onMouseUp: this.handleMouseUp,
                shape: shape,
                type: !href ? type : null,
                level: level
            });

            return React.createElement(
                Element,
                props,
                icon && React.createElement(ButtonIconElement, { value: icon, shape: shape, mini: mini }),
                label,
                children
            );
        }
    }]);

    return Button;
}(Component);

Button.propTypes = {
    children: pt.node,
    disabled: pt.bool,
    flat: pt.bool,
    floating: pt.bool,
    href: pt.string,
    icon: pt.oneOfType([pt.string, pt.element]),
    innerRef: pt.func,
    inverse: pt.bool,
    label: pt.string,
    mini: pt.bool,
    neutral: pt.bool,
    onMouseUp: pt.func,
    onMouseLeave: pt.func,
    primary: pt.bool,
    raised: pt.bool,
    type: pt.string
};
Button.defaultProps = {
    flat: false,
    floating: false,
    innerRef: noop,
    mini: false,
    neutral: true,
    onMouseUp: noop,
    onMouseLeave: noop,
    primary: false,
    raised: false,
    type: 'button'
};
export default Button;
