// mews-ui/lib/Button/style.js
var _templateObject = _taggedTemplateLiteral(['\n    border-radius: ', ';\n    min-width: ', ';\n    padding: ', ';\n\n    & > svg {\n        margin-right: \'5px\';\n    }\n'], ['\n    border-radius: ', ';\n    min-width: ', ';\n    padding: ', ';\n\n    & > svg {\n        margin-right: \'5px\';\n    }\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    &[disabled] {\n        background-color: ', ';\n        box-shadow: ', ';\n    }\n\n    &:active {\n        box-shadow: ', ';\n    }\n\n    &:focus:not(:active) {\n        box-shadow:\n        0 0 8px rgba(0, 0, 0, 0.18),\n        0 8px 16px rgba(0, 0, 0, 0.36);\n    }\n'], ['\n    &[disabled] {\n        background-color: ', ';\n        box-shadow: ', ';\n    }\n\n    &:active {\n        box-shadow: ', ';\n    }\n\n    &:focus:not(:active) {\n        box-shadow:\n        0 0 8px rgba(0, 0, 0, 0.18),\n        0 8px 16px rgba(0, 0, 0, 0.36);\n    }\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    ', '\n\n    ', '\n\n    ', '\n'], ['\n    ', '\n\n    ', '\n\n    ', '\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n        background: ', ';\n        color: ', ';\n    '], ['\n        background: ', ';\n        color: ', ';\n    ']),
    _templateObject5 = _taggedTemplateLiteral(['\n        color: ', ';\n\n        &:focus:not(:active) {\n            background: ', ';\n        }\n    '], ['\n        color: ', ';\n\n        &:focus:not(:active) {\n            background: ', ';\n        }\n    ']),
    _templateObject6 = _taggedTemplateLiteral(['\n        &:hover {\n            background: ', ';\n        }\n    '], ['\n        &:hover {\n            background: ', ';\n        }\n    ']),
    _templateObject7 = _taggedTemplateLiteral(['\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n'], ['\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n        background-color: ', ';\n        color: ', ';\n    '], ['\n        background-color: ', ';\n        color: ', ';\n    ']),
    _templateObject9 = _taggedTemplateLiteral(['\n    align-content: center;\n    align-items: center;\n    border: 0;\n    cursor: pointer;\n    display: inline-block;\n    flex-direction: row;\n    font-family: ', ';\n    font-size: 14px;\n    font-weight: 400;\n    height: ', ';\n    justify-content: center;\n    letter-spacing: 0;\n    line-height: ', ';\n    outline: none;\n    padding: 0;\n    position: relative;\n    text-align: center;\n    text-decoration: none;\n    text-transform: uppercase;\n    transition:\n        box-shadow 0.2s ', ',\n        background-color 0.2s ', ',\n        color 0.2s ', ';\n    white-space: nowrap;\n\n    &::-moz-focus-inner {\n        border: 0;\n    }\n\n    & > * {\n        pointer-events: none;\n    }\n\n    & > input {\n        height: 0.1px;\n        margin: 0;\n        opacity: 0;\n        overflow: hidden;\n        padding: 0;\n        position: absolute;\n        width: 0.1px;\n        z-index: 0;\n    }\n\n    & > svg {\n        display: inline-block;\n        fill: currentColor;\n        font-size: 120%;\n        height: ', ';\n        vertical-align: top;\n        width: 1em;\n    }\n\n    & > .', ' {\n        overflow: hidden;\n    }\n\n    &[disabled] {\n        color: ', ';\n        cursor: auto;\n        pointer-events: none;\n    }\n    \n    ', '\n\n    &:not([disabled]) {\n        ', ';\n        ', ';\n    }\n'], ['\n    align-content: center;\n    align-items: center;\n    border: 0;\n    cursor: pointer;\n    display: inline-block;\n    flex-direction: row;\n    font-family: ', ';\n    font-size: 14px;\n    font-weight: 400;\n    height: ', ';\n    justify-content: center;\n    letter-spacing: 0;\n    line-height: ', ';\n    outline: none;\n    padding: 0;\n    position: relative;\n    text-align: center;\n    text-decoration: none;\n    text-transform: uppercase;\n    transition:\n        box-shadow 0.2s ', ',\n        background-color 0.2s ', ',\n        color 0.2s ', ';\n    white-space: nowrap;\n\n    &::-moz-focus-inner {\n        border: 0;\n    }\n\n    & > * {\n        pointer-events: none;\n    }\n\n    & > input {\n        height: 0.1px;\n        margin: 0;\n        opacity: 0;\n        overflow: hidden;\n        padding: 0;\n        position: absolute;\n        width: 0.1px;\n        z-index: 0;\n    }\n\n    & > svg {\n        display: inline-block;\n        fill: currentColor;\n        font-size: 120%;\n        height: ', ';\n        vertical-align: top;\n        width: 1em;\n    }\n\n    & > .', ' {\n        overflow: hidden;\n    }\n\n    &[disabled] {\n        color: ', ';\n        cursor: auto;\n        pointer-events: none;\n    }\n    \n    ', '\n\n    &:not([disabled]) {\n        ', ';\n        ', ';\n    }\n']),
    _templateObject10 = _taggedTemplateLiteral(['\n                ', ';\n                ', ';\n                box-shadow: ', ';\n            '], ['\n                ', ';\n                ', ';\n                box-shadow: ', ';\n            ']),
    _templateObject11 = _taggedTemplateLiteral(['\n                ', '\n                border-radius: 50%;\n                box-shadow:\n                    0 1px 1.5px 0 rgba(0, 0, 0, 0.12),\n                    0 1px 1px 0 rgba(0, 0, 0, 0.24);\n                font-size: ', ';;\n                height: ', ';\n                width: ', ';\n\n                ', '\n\n                & > .', ' {\n                    border-radius: 50%;\n                }\n            '], ['\n                ', '\n                border-radius: 50%;\n                box-shadow:\n                    0 1px 1.5px 0 rgba(0, 0, 0, 0.12),\n                    0 1px 1px 0 rgba(0, 0, 0, 0.24);\n                font-size: ', ';;\n                height: ', ';\n                width: ', ';\n\n                ', '\n\n                & > .', ' {\n                    border-radius: 50%;\n                }\n            ']),
    _templateObject12 = _taggedTemplateLiteral(['\n                    font-size: ', ';\n                    height: ', ';\n                    width: ', ';\n                '], ['\n                    font-size: ', ';\n                    height: ', ';\n                    width: ', ';\n                ']),
    _templateObject13 = _taggedTemplateLiteral(['\n                ', ';\n                background: transparent;\n            '], ['\n                ', ';\n                background: transparent;\n            ']),
    _templateObject14 = _taggedTemplateLiteral(['', ''], ['', '']),
    _templateObject15 = _taggedTemplateLiteral(['\n    ', '\n'], ['\n    ', '\n']),
    _templateObject16 = _taggedTemplateLiteral(['\n                font-size: 120%;\n                margin-right: ', ';\n                vertical-align: middle;\n            '], ['\n                font-size: 120%;\n                margin-right: ', ';\n                vertical-align: middle;\n            ']),
    _templateObject17 = _taggedTemplateLiteral(['\n                line-height: ', ';\n            '], ['\n                line-height: ', ';\n            ']),
    _templateObject18 = _taggedTemplateLiteral(['\n    ', '\n    background: transparent;\n    border-radius: 50%;\n    vertical-align: middle;\n    width: ', ';\n    height: ', ';\n    padding: 12px;\n\n    box-sizing: border-box;\n\n    & svg {\n        font-size: ', ';\n        line-height: ', ';\n        vertical-align: top;\n    }\n\n    & > .', ' {\n        border-radius: 50%;\n    }\n'], ['\n    ', '\n    background: transparent;\n    border-radius: 50%;\n    vertical-align: middle;\n    width: ', ';\n    height: ', ';\n    padding: 12px;\n\n    box-sizing: border-box;\n\n    & svg {\n        font-size: ', ';\n        line-height: ', ';\n        vertical-align: top;\n    }\n\n    & > .', ' {\n        border-radius: 50%;\n    }\n']),
    _templateObject19 = _taggedTemplateLiteral(['\n    font-size: ', ';\n    vertical-align: top;\n'], ['\n    font-size: ', ';\n    vertical-align: top;\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled, { css } from 'styled-components';
import { alpha } from '../utils/color';
import { ANIMATION_CURVE_DEFAULT, ANIMATION_CURVE_FAST_OUT_LINEAR_IN, PREFERRED_FONT, SHADOW_2P } from '../variables';
import { COLOR_BLACK, COLOR_WHITE, PALETTE_GREY_900 } from '../colors';
import FontIcon from '../FontIcon/FontIcon';
import { RIPPLE_WRAPPER_CLASSNAME } from '../Ripple/style';

var BUTTON_BORDER_RADIUS = '2px';
var BUTTON_HEIGHT = '36px';
var BUTTON_FLOATING_FONT_SIZE = '24px';
var BUTTON_FLOATING_HEIGHT = '56px';
var BUTTON_FLOATING_MINI_HEIGHT = '40px';
var BUTTON_FLOATING_MINI_FONT_SIZE = 'calc(' + BUTTON_FLOATING_MINI_HEIGHT + ' / 2.25)';

var BUTTON_DISABLED_BACKGROUND_COLOR = alpha(COLOR_BLACK, 0.12);
var BUTTON_DISABLED_TEXT_COLOR = alpha(COLOR_BLACK, 0.24);
// const BUTTON_DISABLED_TEXT_COLOR_INVERSE = alpha(COLOR_BLACK, 0.54);
// const BUTTON_DISABLED_BACKGROUND_COLOR_INVERSE = alpha(COLOR_BLACK, 0.08);

var BUTTON_PRIMARY_COLOR = function BUTTON_PRIMARY_COLOR(_ref) {
    var theme = _ref.theme;
    return theme.primaryColor;
};
var BUTTON_PRIMARY_COLOR_HOVER = function BUTTON_PRIMARY_COLOR_HOVER(_ref2) {
    var theme = _ref2.theme;
    return alpha(theme.primaryColor, 0.2);
};
var BUTTON_PRIMARY_COLOR_CONTRAST = function BUTTON_PRIMARY_COLOR_CONTRAST(_ref3) {
    var theme = _ref3.theme;
    return theme.primaryContrastColor;
};

var BUTTON_NEUTRAL_COLOR = COLOR_WHITE;
var BUTTON_NEUTRAL_COLOR_CONTRAST = PALETTE_GREY_900;
var BUTTON_NEUTRAL_COLOR_HOVER = alpha(PALETTE_GREY_900, 0.2);

var BUTTON_SQUARED_ICON_MARGIN = '6px';
var BUTTON_SQUARED_MIN_WIDTH = '88px';
var BUTTON_SQUARED_PADDING = '0 16px';

var ICON_BUTTON_FONT_SIZE = '24px';
var ICON_BUTTON_SIZE = '48px';

var squaredButtonStyles = css(_templateObject, BUTTON_BORDER_RADIUS, BUTTON_SQUARED_MIN_WIDTH, BUTTON_SQUARED_PADDING);

var solidButtonStyles = css(_templateObject2, BUTTON_DISABLED_BACKGROUND_COLOR, SHADOW_2P, SHADOW_2P);

var primaryStyles = css(_templateObject3, function (_ref4) {
    var shape = _ref4.shape;
    return (shape === 'raised' || shape === 'floating') && css(_templateObject4, BUTTON_PRIMARY_COLOR, BUTTON_PRIMARY_COLOR_CONTRAST);
}, function (_ref5) {
    var shape = _ref5.shape;
    return (shape === 'flat' || shape === 'icon') && css(_templateObject5, BUTTON_PRIMARY_COLOR, BUTTON_PRIMARY_COLOR_HOVER);
}, function (_ref6) {
    var shape = _ref6.shape;
    return shape === 'flat' && css(_templateObject6, BUTTON_PRIMARY_COLOR_HOVER);
});

var neutralStyles = css(_templateObject7, function (_ref7) {
    var shape = _ref7.shape;
    return (shape === 'raised' || shape === 'floating') && css(_templateObject8, BUTTON_NEUTRAL_COLOR, BUTTON_NEUTRAL_COLOR_CONTRAST);
}, function (_ref8) {
    var shape = _ref8.shape;
    return (shape === 'flat' || shape === 'icon') && css(_templateObject5, BUTTON_NEUTRAL_COLOR_CONTRAST, BUTTON_NEUTRAL_COLOR_HOVER);
}, function (props) {
    return props.shape === 'icon' && css(_templateObject5, props.theme.iconColor, BUTTON_NEUTRAL_COLOR_HOVER);
}, function (_ref9) {
    var shape = _ref9.shape;
    return shape === 'flat' && css(_templateObject6, BUTTON_NEUTRAL_COLOR_HOVER);
});

// TODO apply reset
var buttonStyles = css(_templateObject9, PREFERRED_FONT, BUTTON_HEIGHT, BUTTON_HEIGHT, ANIMATION_CURVE_FAST_OUT_LINEAR_IN, ANIMATION_CURVE_DEFAULT, ANIMATION_CURVE_DEFAULT, BUTTON_HEIGHT, RIPPLE_WRAPPER_CLASSNAME, BUTTON_DISABLED_TEXT_COLOR, function (_ref10) {
    var shape = _ref10.shape;

    if (shape === 'raised') {
        return css(_templateObject10, squaredButtonStyles, solidButtonStyles, SHADOW_2P);
    } else if (shape === 'floating') {
        return css(_templateObject11, solidButtonStyles, BUTTON_FLOATING_FONT_SIZE, BUTTON_FLOATING_HEIGHT, BUTTON_FLOATING_HEIGHT, function (_ref11) {
            var mini = _ref11.mini;
            return mini && css(_templateObject12, BUTTON_FLOATING_MINI_FONT_SIZE, BUTTON_FLOATING_MINI_HEIGHT, BUTTON_FLOATING_MINI_HEIGHT);
        }, RIPPLE_WRAPPER_CLASSNAME);
    } else if (shape === 'flat') {
        return css(_templateObject13, squaredButtonStyles);
    }
}, function (_ref12) {
    var level = _ref12.level;
    return level === 'primary' && primaryStyles;
}, function (_ref13) {
    var level = _ref13.level;
    return level === 'neutral' && neutralStyles;
});

export var ButtonElement = styled.button(_templateObject14, buttonStyles);
export var ButtonHrefElement = styled.a(_templateObject14, buttonStyles);

export var ButtonIconElement = styled(FontIcon)(_templateObject15, function (_ref14) {
    var shape = _ref14.shape,
        mini = _ref14.mini;

    if (shape === 'raised' || shape === 'flat') {
        return css(_templateObject16, BUTTON_SQUARED_ICON_MARGIN);
    } else if (shape === 'floating') {
        var lineHeight = mini ? BUTTON_FLOATING_MINI_HEIGHT : BUTTON_FLOATING_HEIGHT;
        return css(_templateObject17, lineHeight);
    }
});

var iconButtonStyles = css(_templateObject18, buttonStyles, ICON_BUTTON_SIZE, ICON_BUTTON_SIZE, ICON_BUTTON_FONT_SIZE, BUTTON_HEIGHT, RIPPLE_WRAPPER_CLASSNAME);

export var IconButtonElement = styled.button(_templateObject14, iconButtonStyles);
export var IconButtonHrefElement = styled.a(_templateObject14, iconButtonStyles);

export var IconButtonIconElement = styled(FontIcon)(_templateObject19, ICON_BUTTON_FONT_SIZE);
