// mews-js/src/components/globalization/utils/ensureCultureInfo.js
import _ from 'lodash';
import moment from 'moment-timezone';
import { localeData } from 'moment';
import type from '../../../utils/type';
import CultureInfoComponent from '../cultureInfos/cultureInfoComponent';
import Culture from '../cultures/culture';

export default function ensureCultureInfo(culture, cultureInfoComponent) {
    type.check(cultureInfoComponent, CultureInfoComponent);
    type.check(culture, Culture);

    var cultureInfo = cultureInfoComponent.getByCode(culture.getCode());
    if (_.isNull(cultureInfo)) {
        throw new Error('Missing a locale for the culture: ' + culture.getCode());
    }

    if (_.isNull(localeData(cultureInfo.getCode()))) {
        defineMomentLocale(cultureInfo);
    }

    return cultureInfo;
}

function defineMomentLocale(cultureInfo) {
    const dateTime = cultureInfo.getDateTimeInfo();
    const [am, pm] = dateTime.meridiem;

    moment.defineLocale(cultureInfo.getCode(), {
        months: dateTime.monthNames,
        monthsShort: dateTime.monthShortNames,
        weekdays: dateTime.dayNames,
        weekdaysShort: dateTime.dayShortNames,
        weekdaysMin: dateTime.dayMinNames,
        longDateFormat: {
            LT: dateTime.shortTimeFormat,
            LTS: dateTime.timeFormat,
            L: dateTime.dateFormat,
            LL: dateTime.longDateFormat,
            LLL: dateTime.dateTimeFormat,
        },
        week: {
            dow: dateTime.firstDayOfWeek,
        },
        meridiemParse: new RegExp(`${am}|${pm}`),
        isPM: value => value === pm,
    });
}
