// mews-ui/lib/Radio/Radio.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import { PropTypes as pt } from 'prop-types';
import { noop } from '../utils/func';
import { RadioElement, RadioInputElement, RadioBackgroundElement, RadioOuterCircleElement, RadioInnerCircleElement } from './style';

var Radio = function (_Component) {
    _inherits(Radio, _Component);

    function Radio() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, Radio);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Radio.__proto__ || Object.getPrototypeOf(Radio)).call.apply(_ref, [this].concat(args))), _this), _this.handleChange = function (e) {
            if (!_this.props.disabled) {
                _this.props.onChange(!_this.props.checked, e);
            }
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Radio, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                className = _props.className,
                checked = _props.checked,
                disabled = _props.disabled,
                other = _objectWithoutProperties(_props, ['className', 'checked', 'disabled']);

            return React.createElement(
                RadioElement,
                { className: className },
                React.createElement(RadioInputElement, _extends({}, other, { checked: checked, disabled: disabled, onChange: this.handleChange })),
                React.createElement(
                    RadioBackgroundElement,
                    null,
                    React.createElement(RadioOuterCircleElement, { checked: checked, disabled: disabled }),
                    React.createElement(RadioInnerCircleElement, { checked: checked, disabled: disabled })
                )
            );
        }
    }]);

    return Radio;
}(Component);

Radio.propTypes = {
    className: pt.string,
    checked: pt.bool,
    disabled: pt.bool,
    onChange: pt.func
};
Radio.defaultProps = {
    onChange: noop
};
export default Radio;
