// mews-ui/lib/Radio/style.js
var _templateObject = _taggedTemplateLiteral([''], ['']),
    _templateObject2 = _taggedTemplateLiteral(['\n    display: inline-flex;\n    align-items: center;    \n'], ['\n    display: inline-flex;\n    align-items: center;    \n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    display: inline-block;\n    position: relative;\n    box-sizing: border-box;\n    width: ', ';\n    height: ', ';\n    padding: 10px;\n    cursor: pointer;\n'], ['\n    display: inline-block;\n    position: relative;\n    box-sizing: border-box;\n    width: ', ';\n    height: ', ';\n    padding: 10px;\n    cursor: pointer;\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n    position: absolute;\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n    margin: 0;\n    padding: 0;\n\n    opacity: 0;\n    z-index: 1;\n'], ['\n    position: absolute;\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n    margin: 0;\n    padding: 0;\n\n    opacity: 0;\n    z-index: 1;\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n    display: inline-block;\n    position: absolute;\n    left: 10px;\n    width: 50%;\n    height: 50%;\n    box-sizing: border-box;\n'], ['\n    display: inline-block;\n    position: absolute;\n    left: 10px;\n    width: 50%;\n    height: 50%;\n    box-sizing: border-box;\n']),
    _templateObject6 = _taggedTemplateLiteral(['\n    position: absolute;\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n    border-width: 2px;\n    border-style: solid;\n    border-radius: 50%;\n    box-sizing: border-box;\n    border-color: ', ';\n    transition: border-color 120ms ', ';\n\n    ', '\n'], ['\n    position: absolute;\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n    border-width: 2px;\n    border-style: solid;\n    border-radius: 50%;\n    box-sizing: border-box;\n    border-color: ', ';\n    transition: border-color 120ms ', ';\n\n    ', '\n']),
    _templateObject7 = _taggedTemplateLiteral(['\n    position: absolute;\n    left: 0;\n    top: 0;\n    width: 100%;\n    height: 100%;\n    border-radius: 50%;\n    background-color: ', ';\n    transform: scale(0, 0);\n    transition: transform 120ms ', ', background-color 120ms ', ';\n\n    ', '\n'], ['\n    position: absolute;\n    left: 0;\n    top: 0;\n    width: 100%;\n    height: 100%;\n    border-radius: 50%;\n    background-color: ', ';\n    transform: scale(0, 0);\n    transition: transform 120ms ', ', background-color 120ms ', ';\n\n    ', '\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled from 'styled-components';
import { COLOR_BLACK } from '../colors';
import { DEFAULT_THEME, ANIMATION_CURVE_DEFAULT } from '../variables';
import { alpha } from '../utils/color';

var RADIO_SIZE = '40px';
var DISABLED_COLOR = alpha(COLOR_BLACK, 0.24);

export var RadioGroupElement = styled.div(_templateObject);

export var RadioButtonElement = styled.label(_templateObject2);

export var RadioLabelElement = styled.span(_templateObject);

export var RadioElement = styled.div(_templateObject3, RADIO_SIZE, RADIO_SIZE);

export var RadioInputElement = styled.input.attrs({
    type: 'radio'
})(_templateObject4);

export var RadioBackgroundElement = styled.div(_templateObject5);

export var RadioOuterCircleElement = styled.div(_templateObject6, function (props) {
    return props.disabled ? DISABLED_COLOR : props.theme.secondaryTextColor;
}, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.checked && '\n        border-color: ' + (props.disabled ? DISABLED_COLOR : props.theme.primaryColor) + ';\n    ';
});
RadioOuterCircleElement.defaultProps = { theme: DEFAULT_THEME };

export var RadioInnerCircleElement = styled.div(_templateObject7, function (props) {
    return props.disabled ? DISABLED_COLOR : props.theme.secondaryTextColor;
}, ANIMATION_CURVE_DEFAULT, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.checked && '\n        background-color: ' + (props.disabled ? DISABLED_COLOR : props.theme.primaryColor) + ';\n        transform: scale(0.5);\n    ';
});
RadioInnerCircleElement.defaultProps = { theme: DEFAULT_THEME };
