// mews-js/src/utils/check.js
import moment from 'moment-timezone';
import _ from 'lodash';

const BANNED_TIMEZONES = [
    'etc/utc',
    'etc/universal',
    'etc/zulu',
    'utc',
    'universal',
    'zulu',
    'etc/uct',
    'uct',
];

export default {
    notUtcTimezone(timezone) {
        const normalizedTimezone = _.toLower(timezone);
        if (_.some(BANNED_TIMEZONES, t => t === normalizedTimezone)) {
            throw new Error('UTC timezone is not allowed.');
        }
    },

    timezone(timezone) {
        if (_.isNull(moment.tz.zone(timezone))) {
            throw new Error('Unknown timezone identifier: ' + timezone);
        }
    },

    string(string) {
        if (!_.isString(string)) {
            throw new TypeError('Expected to be a string.');
        }
    },

    number(number) {
        if (!_.isNumber(number)) {
            throw new TypeError('Expected a number: "' + number + '"');
        }
    },

    data(object) {
        if (!_.isObject(object)) {
            throw new TypeError('Expected the data to be an object.');
        }
    },

    sameTimezones(entity1, entity2) {
        if (entity1.timezone !== entity2.timezone) {
            throw new Error('Entities are in different timezones.');
        }
    },
};
