// mews-js/src/components/sentryComponent.js
import Raven from 'raven-js';
import { type } from '../utils';

export default function SentryComponent(configurationComponent) {
    Raven
        .config(
            configurationComponent.getSentryDsn(),
            getConfiguration()
        )
        .install();

    return type.create(SentryComponent, {
        logError(error, extra = {}) {
            if (Raven.isSetup()) {
                if (error instanceof Error) {
                    Raven.captureException(error, { extra });
                } else {
                    Raven.captureMessage(error, { extra });
                }
            }
        },

        setPerson(person) {
            if (Raven.isSetup()) {
                Raven.setUserContext(person);
            }
        },
    });

    function getConfiguration() {
        return {
            environment: configurationComponent.getPlatformMode(),
            whitelistUrls: configurationComponent.getPlatform().urls,
            release: configurationComponent.getVersionHash(),
            tags: {
                product: configurationComponent.getProductName(),
            }
        };
    }
}
