// mews-js/src/components/globalization/componentEnhancers/comparable.js
import { check } from '../../../utils';

export default function comparable(toMoment) {
    return {
        isEqual(date1, date2) {
            check.sameTimezones(date1, date2);
            return toMoment(date1).isSame(toMoment(date2));
        },

        isLess(date1, date2) {
            check.sameTimezones(date1, date2);
            return toMoment(date1).isBefore(toMoment(date2));
        },

        isLessOrEqual,

        isGreater(date1, date2) {
            check.sameTimezones(date1, date2);
            return toMoment(date1).isAfter(toMoment(date2));
        },

        isGreaterOrEqual,

        min(date1, date2) {
            return isLessOrEqual(date1, date2) ? date1 : date2;
        },

        max(date1, date2) {
            return isGreaterOrEqual(date1, date2) ? date1 : date2;
        }
    };

    function isLessOrEqual(date1, date2) {
        check.sameTimezones(date1, date2);
        return toMoment(date1).isSameOrBefore(toMoment(date2));
    }

    function isGreaterOrEqual(date1, date2) {
        check.sameTimezones(date1, date2);
        return toMoment(date1).isSameOrAfter(toMoment(date2));
    }
}
