// mews-ui/lib/Sidebar/style.js
var _templateObject = _taggedTemplateLiteral(['\n    min-width: 320px;\n    position: fixed;\n    top: 0;\n    left: 0;\n    bottom: 0;\n    right: 0;\n    overflow: hidden;\n    pointer-events: none;\n    z-index: ', ';\n'], ['\n    min-width: 320px;\n    position: fixed;\n    top: 0;\n    left: 0;\n    bottom: 0;\n    right: 0;\n    overflow: hidden;\n    pointer-events: none;\n    z-index: ', ';\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    position: absolute;\n    width: 100%;\n    height: 100%;\n    background-color: ', '; \n    transition: opacity ', ' ', ';\n    opacity: 0;\n\n    ', '\n    ', '\n\n    ', '\n    ', '\n\n    ', ';\n'], ['\n    position: absolute;\n    width: 100%;\n    height: 100%;\n    background-color: ', '; \n    transition: opacity ', ' ', ';\n    opacity: 0;\n\n    ', '\n    ', '\n\n    ', '\n    ', '\n\n    ', ';\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n        pointer-events: initial;\n    '], ['\n        pointer-events: initial;\n    ']),
    _templateObject4 = _taggedTemplateLiteral(['\n    min-width: 320px;\n    width: ', ';\n    height: 100%;\n    position: absolute;\n    top: 0;\n    right: 0;\n    pointer-events: initial;\n    display: flex;\n    flex-direction: column;\n    background-color: ', ';\n    transition: transform ', ' ', ';\n\n    ', '\n    ', '\n\n    ', '\n    ', '\n\n    ', '\n'], ['\n    min-width: 320px;\n    width: ', ';\n    height: 100%;\n    position: absolute;\n    top: 0;\n    right: 0;\n    pointer-events: initial;\n    display: flex;\n    flex-direction: column;\n    background-color: ', ';\n    transition: transform ', ' ', ';\n\n    ', '\n    ', '\n\n    ', '\n    ', '\n\n    ', '\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n        box-shadow: ', ';\n    '], ['\n        box-shadow: ', ';\n    ']),
    _templateObject6 = _taggedTemplateLiteral(['\n    display: flex;\n    align-items: center;\n    flex-shrink: 0;\n    border-bottom: 1px solid;\n    border-color: ', ';\n    box-sizing: border-box;\n    padding-right: 16px;\n'], ['\n    display: flex;\n    align-items: center;\n    flex-shrink: 0;\n    border-bottom: 1px solid;\n    border-color: ', ';\n    box-sizing: border-box;\n    padding-right: 16px;\n']),
    _templateObject7 = _taggedTemplateLiteral(['\n    white-space: nowrap;\n    text-overflow: ellipsis;\n'], ['\n    white-space: nowrap;\n    text-overflow: ellipsis;\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n    flex-grow: 1;\n    overflow-y: auto;\n    overflow-x: hidden;\n    padding: 16px;\n'], ['\n    flex-grow: 1;\n    overflow-y: auto;\n    overflow-x: hidden;\n    padding: 16px;\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled, { css } from 'styled-components';
import { PALETTE_GREY_900, COLOR_WHITE } from '../colors';
import { ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, Z_DEPTH_SHADOW_2, DEFAULT_THEME, Z_INDEX_HIGH } from '../variables';
import StyledTransitionGroup from '../TransitionGroup/StyledTransitionGroup';

export var SidebarContainerElement = styled(StyledTransitionGroup)(_templateObject, Z_INDEX_HIGH);

export var BackdropElement = styled.div(_templateObject2, PALETTE_GREY_900, ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.enter && 'opacity: 0;';
}, function (props) {
    return props.enterActive && 'opacity: .48;';
}, function (props) {
    return props.leave && 'opacity: .48;';
}, function (props) {
    return props.leaveActive && 'opacity: 0;';
}, function (props) {
    return props.opened && css(_templateObject3);
});

export var SidebarElement = styled.div(_templateObject4, function (props) {
    return props.width;
}, COLOR_WHITE, ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.enter && 'transform: translateX(100%);';
}, function (props) {
    return props.enterActive && 'transform: translateX(0);';
}, function (props) {
    return props.leave && 'transform: translateX(0);';
}, function (props) {
    return props.leaveActive && 'transform: translateX(100%);';
}, function (props) {
    return props.opened && css(_templateObject5, Z_DEPTH_SHADOW_2);
});

export var SidebarHeaderElement = styled.div(_templateObject6, function (props) {
    return props.theme.dividerColor;
});
SidebarHeaderElement.defaultProps = { theme: DEFAULT_THEME };

export var SidebarHeaderTitleElement = styled.div(_templateObject7);

export var SidebarContentElement = styled.div(_templateObject8);
