// mews-ui/lib/Field/style.js
var _templateObject = _taggedTemplateLiteral(['\n    position: relative;\n    padding-top: 16px;\n\n    ', '\n'], ['\n    position: relative;\n    padding-top: 16px;\n\n    ', '\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    display: flex;\n'], ['\n    display: flex;\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    align-self: center;\n    box-sizing: border-box;\n    margin-bottom: 8px;\n    max-width: ', ';\n    position: relative;\n    padding-bottom: 4px;\n    padding-top: 16px;\n    width: 100%;\n\n    ', '\n\n    border-bottom: 1px solid ', ';\n    ', '\n    ', '\n\n    ', '\n'], ['\n    align-self: center;\n    box-sizing: border-box;\n    margin-bottom: 8px;\n    max-width: ', ';\n    position: relative;\n    padding-bottom: 4px;\n    padding-top: 16px;\n    width: 100%;\n\n    ', '\n\n    border-bottom: 1px solid ', ';\n    ', '\n    ', '\n\n    ', '\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n        border-color: ', ';\n    '], ['\n        border-color: ', ';\n    ']),
    _templateObject5 = _taggedTemplateLiteral(['\n    align-self: center;\n    color: ', ';\n    display: block;\n    font-size: ', ' !important;\n    height: ', ';\n    line-height: ', ' !important;\n    text-align: center;\n    transition: color ', ' ', ';\n    width: ', ';\n    min-width: ', ';\n\n    ', '\n'], ['\n    align-self: center;\n    color: ', ';\n    display: block;\n    font-size: ', ' !important;\n    height: ', ';\n    line-height: ', ' !important;\n    text-align: center;\n    transition: color ', ' ', ';\n    width: ', ';\n    min-width: ', ';\n\n    ', '\n']),
    _templateObject6 = _taggedTemplateLiteral(['\n        color: ', ';\n    '], ['\n        color: ', ';\n    ']),
    _templateObject7 = _taggedTemplateLiteral(['\n    color: ', ';\n    font-size: ', ';\n    font-weight: normal;\n    left: 0;\n    right: 0;\n    overflow: hidden;\n    pointer-events: none;\n    position: absolute;\n    text-overflow: ellipsis;\n    transform-origin: left top;\n    transition-duration: ', ';\n    transition-property: transform, color, font-size, height;\n    transition-timing-function: ', ';\n    white-space: nowrap;\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n'], ['\n    color: ', ';\n    font-size: ', ';\n    font-weight: normal;\n    left: 0;\n    right: 0;\n    overflow: hidden;\n    pointer-events: none;\n    position: absolute;\n    text-overflow: ellipsis;\n    transform-origin: left top;\n    transition-duration: ', ';\n    transition-property: transform, color, font-size, height;\n    transition-timing-function: ', ';\n    white-space: nowrap;\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n        display: none;\n    '], ['\n        display: none;\n    ']),
    _templateObject9 = _taggedTemplateLiteral(['\n        font-size: calc(1em * 0.75);\n        transform: translateY(-100%);\n    '], ['\n        font-size: calc(1em * 0.75);\n        transform: translateY(-100%);\n    ']),
    _templateObject10 = _taggedTemplateLiteral(['\n    color: ', ';\n'], ['\n    color: ', ';\n']),
    _templateObject11 = _taggedTemplateLiteral(['\n    color: ', ';\n    font-size: 12px;\n    padding-bottom: 8px;\n\n    ', '\n'], ['\n    color: ', ';\n    font-size: 12px;\n    padding-bottom: 8px;\n\n    ', '\n']),
    _templateObject12 = _taggedTemplateLiteral(['\n        margin-left: ', 'px;\n    '], ['\n        margin-left: ', 'px;\n    ']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled, { css } from 'styled-components';

import { COLOR_RED } from '../colors';
import { ANIMATION_CURVE_DEFAULT, ANIMATION_DURATION, DEFAULT_THEME } from '../variables';
import FontIcon from '../FontIcon/FontIcon';

var FIELD_TEXT_REQUIRED_COLOR = COLOR_RED;
var FIELD_TEXT_ERROR_COLOR = COLOR_RED;
var FIELD_ICON_FONT_SIZE = '24px';
var FIELD_ICON_SIZE = 'calc(2 * ' + FIELD_ICON_FONT_SIZE + ')';

export var FIELD_TEXT_NORMAL_SIZE = '16px';
export var FIELD_TEXT_LARGE_SIZE = '25px';

export var FieldElement = styled.div(_templateObject, function (props) {
    return props.hidden && '\n        display: none;\n    ';
});

export var ContentWrapperElement = styled.div(_templateObject2);

export var InputWrapperElement = styled.div(_templateObject3, function (props) {
    return props.icon ? 'calc(100% - ' + FIELD_ICON_SIZE + ')' : '100%';
}, function (props) {
    return props.boxed && '\n        padding-left: 16px;\n        padding-right: 16px;\n    ';
}, function (props) {
    return props.theme.hintTextColor;
}, function (props) {
    return props.focused && css(_templateObject4, props.theme.primaryColor);
}, function (props) {
    return !props.focused && props.error && css(_templateObject4, FIELD_TEXT_ERROR_COLOR);
}, function (props) {
    return props.boxed && '\n        background-color: rgba(0, 0, 0, .06);\n        border-radius: 4px;\n    ';
});
InputWrapperElement.defaultProps = { theme: DEFAULT_THEME };

export var IconElement = styled(FontIcon)(_templateObject5, function (props) {
    return props.theme.iconColor;
}, FIELD_ICON_FONT_SIZE, FIELD_ICON_SIZE, FIELD_ICON_SIZE, ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, FIELD_ICON_SIZE, FIELD_ICON_SIZE, function (props) {
    return props.focused && css(_templateObject6, props.theme.primaryColor);
});
IconElement.defaultProps = { theme: DEFAULT_THEME };

export var LabelElement = styled.label(_templateObject7, function (props) {
    return props.theme.hintTextColor;
}, function (props) {
    return props.large ? FIELD_TEXT_LARGE_SIZE : FIELD_TEXT_NORMAL_SIZE;
}, ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.boxed && '\n        left: 16px;\n        right: 16px;\n    ';
}, function (props) {
    return !props.floating && (props.focused || props.filled) && css(_templateObject8);
}, function (props) {
    return props.floating && (props.focused || props.filled) && css(_templateObject9);
}, function (props) {
    return props.focused && css(_templateObject6, props.theme.primaryColor);
}, function (props) {
    return !props.focused && props.error && css(_templateObject6, FIELD_TEXT_REQUIRED_COLOR);
});
LabelElement.defaultProps = { theme: DEFAULT_THEME };

export var LabelRequiredElement = styled.span(_templateObject10, FIELD_TEXT_REQUIRED_COLOR);

export var ErrorElement = styled.div(_templateObject11, FIELD_TEXT_ERROR_COLOR, function (props) {
    return props.icon && css(_templateObject12, props.boxed ? 64 : 48);
});
