// mews-ui/lib/Dialog/style.js
var _templateObject = _taggedTemplateLiteral(['\n    position: fixed;\n    top: 0;\n    left: 0;\n    bottom: 0;\n    right: 0;\n    overflow: hidden;\n\n    display: flex;\n    align-items: center;\n    justify-content: center;\n\n    pointer-events: none;\n    z-index: ', ';\n'], ['\n    position: fixed;\n    top: 0;\n    left: 0;\n    bottom: 0;\n    right: 0;\n    overflow: hidden;\n\n    display: flex;\n    align-items: center;\n    justify-content: center;\n\n    pointer-events: none;\n    z-index: ', ';\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    position: absolute;\n    width: 100%;\n    height: 100%;\n    background-color: ', ';\n    transition: opacity ', ' ', ';\n\n    pointer-events: initial;\n\n    ', '\n    ', '\n\n    ', '\n    ', '\n'], ['\n    position: absolute;\n    width: 100%;\n    height: 100%;\n    background-color: ', ';\n    transition: opacity ', ' ', ';\n\n    pointer-events: initial;\n\n    ', '\n    ', '\n\n    ', '\n    ', '\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    background-color: ', ';\n    box-shadow: ', ';\n    z-index: 1;\n    width: 50vw;\n    max-width: 98vw;\n    max-height: 98vh;\n\n    pointer-events: initial;\n\n    transition-duration: ', ';\n    transition-timing-function: ', ';\n    transition-property: opacity, transform;\n\n    ', '\n    ', '\n\n    ', '\n    ', '\n'], ['\n    background-color: ', ';\n    box-shadow: ', ';\n    z-index: 1;\n    width: 50vw;\n    max-width: 98vw;\n    max-height: 98vh;\n\n    pointer-events: initial;\n\n    transition-duration: ', ';\n    transition-timing-function: ', ';\n    transition-property: opacity, transform;\n\n    ', '\n    ', '\n\n    ', '\n    ', '\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n    padding: 24px;\n'], ['\n    padding: 24px;\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n    padding: 8px;\n    text-align: right;\n'], ['\n    padding: 8px;\n    text-align: right;\n']),
    _templateObject6 = _taggedTemplateLiteral(['\n    &:not(:last-child) {\n        margin-right: 8px;\n    }\n'], ['\n    &:not(:last-child) {\n        margin-right: 8px;\n    }\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled from 'styled-components';
import { PALETTE_GREY_900, COLOR_WHITE } from '../colors';
import { ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, Z_DEPTH_SHADOW_3, Z_INDEX_HIGH } from '../variables';
import StyledTransitionGroup from '../TransitionGroup/StyledTransitionGroup';
import Button from '../Button/Button';

export var DialogContainerElement = styled(StyledTransitionGroup)(_templateObject, Z_INDEX_HIGH);

export var BackdropElement = styled.div(_templateObject2, PALETTE_GREY_900, ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.enter && 'opacity: 0;';
}, function (props) {
    return props.enterActive && 'opacity: .48;';
}, function (props) {
    return props.leave && 'opacity: .48;';
}, function (props) {
    return props.leaveActive && 'opacity: 0;';
});

export var DialogElement = styled.div(_templateObject3, COLOR_WHITE, Z_DEPTH_SHADOW_3, ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.enter && '\n        opacity: 0;\n        transform: translateY(-4em);\n    ';
}, function (props) {
    return props.enterActive && '\n        opacity: 1;\n        transform: translateY(0);\n   ';
}, function (props) {
    return props.leave && '\n        opacity: 1;\n        transform: translateY(0);\n    ';
}, function (props) {
    return props.leaveActive && '\n        opacity: 0;\n        transform: translateY(-4em);\n    ';
});

export var DialogContentElement = styled.div(_templateObject4);

export var DialogActionsElement = styled.div(_templateObject5);

export var DialogButton = styled(Button)(_templateObject6);
