// mews-ui/src/Radio/Radio.jsx
import React, { Component } from 'react';
import { PropTypes as pt } from 'prop-types';
import { noop } from '../utils/func';
import { RadioElement, RadioInputElement, RadioBackgroundElement, RadioOuterCircleElement, RadioInnerCircleElement } from './style';

export default class Radio extends Component {
    static propTypes = {
        className: pt.string,
        checked: pt.bool,
        disabled: pt.bool,
        onChange: pt.func,
    };

    static defaultProps = {
        onChange: noop,
    };

    handleChange = e => {
        if (!this.props.disabled) {
            this.props.onChange(!this.props.checked, e);
        }
    };

    render() {
        const { className, checked, disabled, ...other } = this.props;

        return (
            <RadioElement className={className}>
                <RadioInputElement {...other} checked={checked} disabled={disabled} onChange={this.handleChange} />
                <RadioBackgroundElement>
                    <RadioOuterCircleElement checked={checked} disabled={disabled} />
                    <RadioInnerCircleElement checked={checked} disabled={disabled} />
                </RadioBackgroundElement>
            </RadioElement>
        );
    }
}
