// mews-ui/src/Radio/style.jsx
import styled from 'styled-components';
import { COLOR_BLACK } from '../colors';
import { DEFAULT_THEME, ANIMATION_CURVE_DEFAULT } from '../variables';
import { alpha } from '../utils/color';

const RADIO_SIZE = '40px';
const DISABLED_COLOR = alpha(COLOR_BLACK, 0.24);

export const RadioGroupElement = styled.div``;

export const RadioButtonElement = styled.label`
    display: inline-flex;
    align-items: center;    
`;

export const RadioLabelElement = styled.span``;

export const RadioElement = styled.div`
    display: inline-block;
    position: relative;
    box-sizing: border-box;
    width: ${RADIO_SIZE};
    height: ${RADIO_SIZE};
    padding: 10px;
    cursor: pointer;
`;

export const RadioInputElement = styled.input.attrs({
    type: 'radio',
})`
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0;

    opacity: 0;
    z-index: 1;
`;

export const RadioBackgroundElement = styled.div`
    display: inline-block;
    position: absolute;
    left: 10px;
    width: 50%;
    height: 50%;
    box-sizing: border-box;
`;

export const RadioOuterCircleElement = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border-width: 2px;
    border-style: solid;
    border-radius: 50%;
    box-sizing: border-box;
    border-color: ${props => (props.disabled ? DISABLED_COLOR : props.theme.secondaryTextColor)};
    transition: border-color 120ms ${ANIMATION_CURVE_DEFAULT};

    ${props => props.checked && `
        border-color: ${props.disabled ? DISABLED_COLOR : props.theme.primaryColor};
    `}
`;
RadioOuterCircleElement.defaultProps = { theme: DEFAULT_THEME };

export const RadioInnerCircleElement = styled.div`
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    border-radius: 50%;
    background-color: ${props => (props.disabled ? DISABLED_COLOR : props.theme.secondaryTextColor)};
    transform: scale(0, 0);
    transition: transform 120ms ${ANIMATION_CURVE_DEFAULT}, background-color 120ms ${ANIMATION_CURVE_DEFAULT};

    ${props => props.checked && `
        background-color: ${props.disabled ? DISABLED_COLOR : props.theme.primaryColor};
        transform: scale(0.5);
    `}
`;
RadioInnerCircleElement.defaultProps = { theme: DEFAULT_THEME };
