// mews-ui/lib/TimePicker/style.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _templateObject = _taggedTemplateLiteral(['\n    padding: 8px;\n    box-sizing: border-box;\n\n    display: flex;\n'], ['\n    padding: 8px;\n    box-sizing: border-box;\n\n    display: flex;\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    margin-left: calc(25% - 16px);\n'], ['\n    margin-left: calc(25% - 16px);\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    display: flex;\n    align-items: center;\n'], ['\n    display: flex;\n    align-items: center;\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n    min-width: 25%;\n'], ['\n    min-width: 25%;\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n    width: 48px;\n    margin-left: 16px;\n    border-bottom: 1px solid ', ';\n    padding-bottom: 4px;\n\n    &:focus {\n        border-color: ', ';\n    }\n'], ['\n    width: 48px;\n    margin-left: 16px;\n    border-bottom: 1px solid ', ';\n    padding-bottom: 4px;\n\n    &:focus {\n        border-color: ', ';\n    }\n']),
    _templateObject6 = _taggedTemplateLiteral(['\n    width: 64px;\n    height: ', 'px;\n    overflow: hidden;\n    position: relative;\n    touch-action: none;\n'], ['\n    width: 64px;\n    height: ', 'px;\n    overflow: hidden;\n    position: relative;\n    touch-action: none;\n']),
    _templateObject7 = _taggedTemplateLiteral(['\n    width: 100%;\n    height: 100%;\n    position: absolute;\n    left: 0;\n    top: ', 'px;\n    transition: transform 200ms ', ';\n'], ['\n    width: 100%;\n    height: 100%;\n    position: absolute;\n    left: 0;\n    top: ', 'px;\n    transition: transform 200ms ', ';\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n    position: absolute;\n    left: 0;\n    top: ', 'px;\n    background-color: ', ';\n    height: ', 'px;\n    width: 100%;\n    box-sizing: border-box;\n    pointer-events: none;\n    display: flex;\n    align-items: center;\n    justify-content: flex-end;\n    font-weight: 500;\n    color: #fff;\n'], ['\n    position: absolute;\n    left: 0;\n    top: ', 'px;\n    background-color: ', ';\n    height: ', 'px;\n    width: 100%;\n    box-sizing: border-box;\n    pointer-events: none;\n    display: flex;\n    align-items: center;\n    justify-content: flex-end;\n    font-weight: 500;\n    color: #fff;\n']),
    _templateObject9 = _taggedTemplateLiteral(['\n    font-size: 14px;\n    width: 100%;\n    height: ', 'px;\n    display: flex;\n    align-items: center;\n    justify-content: center;\n    cursor: pointer;\n    box-sizing: border-box;\n    user-select: none;\n\n    ', '\n'], ['\n    font-size: 14px;\n    width: 100%;\n    height: ', 'px;\n    display: flex;\n    align-items: center;\n    justify-content: center;\n    cursor: pointer;\n    box-sizing: border-box;\n    user-select: none;\n\n    ', '\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled from 'styled-components';
import { ANIMATION_CURVE_DEFAULT, DEFAULT_THEME } from '../variables';
import Slider from '../Slider/Slider';
import RadioGroup from '../Radio/RadioGroup';
import RadioButton from '../Radio/RadioButton';
import { InputElement as x } from '../Input/style';

export var ITEM_HEIGHT = 48;

export var TimePickerElement = styled.div(_templateObject);

export var MeridiemSelectorElement = styled(RadioGroup)(_templateObject2);

export var MeridiemButtonElement = RadioButton;

export var NumberSelectorElement = styled.div(_templateObject3);

export var LabelElement = styled.div(_templateObject4);

export var SliderElement = Slider;
SliderElement.defaultProps = _extends({}, SliderElement.defaultProps, {
    step: 1
});

export var InputElement = styled(x).attrs({
    type: 'number',
    step: 1
})(_templateObject5, function (props) {
    return props.theme.hintTextColor;
}, function (props) {
    return props.theme.primaryColor;
});

InputElement.defaultProps = _extends({}, InputElement.defaultProps, {
    theme: DEFAULT_THEME
});

export var ScrollableSelectorElement = styled.div(_templateObject6, ITEM_HEIGHT * 5);

export var ScrollContainerElement = styled.div(_templateObject7, ITEM_HEIGHT * 2, ANIMATION_CURVE_DEFAULT);

export var ScrollItemHighlightElement = styled.div(_templateObject8, ITEM_HEIGHT * 2, function (props) {
    return props.theme.primaryColor;
}, ITEM_HEIGHT);
ScrollItemHighlightElement.defaultProps = { theme: DEFAULT_THEME };

export var ScrollItemElement = styled.div.attrs({
    onClick: function onClick(props) {
        return function () {
            return props.onItemClick(props.value);
        };
    }
})(_templateObject9, ITEM_HEIGHT, function (props) {
    return props.selected && '\n        font-weight: 500;\n        color: #fff;\n    ';
});
