// mews-js/lib/components/globalization/componentEnhancers/comparable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = comparable;

var _utils = require('../../../utils');

function comparable(toMoment) {
    return {
        isEqual: function isEqual(date1, date2) {
            _utils.check.sameTimezones(date1, date2);
            return toMoment(date1).isSame(toMoment(date2));
        },
        isLess: function isLess(date1, date2) {
            _utils.check.sameTimezones(date1, date2);
            return toMoment(date1).isBefore(toMoment(date2));
        },


        isLessOrEqual: isLessOrEqual,

        isGreater: function isGreater(date1, date2) {
            _utils.check.sameTimezones(date1, date2);
            return toMoment(date1).isAfter(toMoment(date2));
        },


        isGreaterOrEqual: isGreaterOrEqual,

        min: function min(date1, date2) {
            return isLessOrEqual(date1, date2) ? date1 : date2;
        },
        max: function max(date1, date2) {
            return isGreaterOrEqual(date1, date2) ? date1 : date2;
        }
    };

    function isLessOrEqual(date1, date2) {
        _utils.check.sameTimezones(date1, date2);
        return toMoment(date1).isSameOrBefore(toMoment(date2));
    }

    function isGreaterOrEqual(date1, date2) {
        _utils.check.sameTimezones(date1, date2);
        return toMoment(date1).isSameOrAfter(toMoment(date2));
    }
}
