// mews-js/src/react/components/localization/LocalizedCurrencyValue.js
import _ from 'lodash';
import React, { Component } from 'react';
import { PropTypes as pt } from 'prop-types';
import { localizationPropTypes } from './propTypes';

class LocalizedCurrencyValue extends Component {
    static propTypes = {
        valueCheck: pt.func
    }

    static contextTypes = {
        localization: localizationPropTypes,
    }

    componentDidMount() {
        this.unsubscribe = this.context.localization.subscribe(() => this.forceUpdate());
    }

    componentWillUnmount() {
        this.unsubscribe();
    }

    render() {
        const { value, valueCheck, ...other } = this.props;

        let finalValue = value;
        if (valueCheck) {
            finalValue = valueCheck(finalValue);
        }
        if (_.isNil(finalValue) || !valueCheck) {
            finalValue = this.context.localization.localizeCurrencyValue(value);
        }

        return (
            <span {...other}>{finalValue}</span>
        );
    }
}

export default LocalizedCurrencyValue;
