// mews-js/lib/components/sentryComponent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = SentryComponent;

var _ravenJs = require('raven-js');

var _ravenJs2 = _interopRequireDefault(_ravenJs);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function SentryComponent(configurationComponent) {
    _ravenJs2.default.config(configurationComponent.getSentryDsn(), getConfiguration()).install();

    return _utils.type.create(SentryComponent, {
        logError: function logError(error) {
            var extra = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

            if (_ravenJs2.default.isSetup()) {
                if (error instanceof Error) {
                    _ravenJs2.default.captureException(error, { extra: extra });
                } else {
                    _ravenJs2.default.captureMessage(error, { extra: extra });
                }
            }
        },
        setPerson: function setPerson(person) {
            if (_ravenJs2.default.isSetup()) {
                _ravenJs2.default.setUserContext(person);
            }
        }
    });

    function getConfiguration() {
        return {
            environment: configurationComponent.getPlatformMode(),
            whitelistUrls: configurationComponent.getPlatform().urls,
            release: configurationComponent.getVersionHash(),
            tags: {
                product: configurationComponent.getProductName()
            }
        };
    }
}
