// mews-ui/src/Avatar/style.jsx
import styled, { css } from 'styled-components';
import FontIcon from '../FontIcon/FontIcon';

import { COLOR_WHITE, PALETTE_GREY_500 } from '../colors';

const AVATAR_COLOR = COLOR_WHITE;
const AVATAR_BACKGROUND = PALETTE_GREY_500;
const AVATAR_SIZE = '40px';
const AVATAR_FONT_SIZE = '24px';

// todo apply reset
export const AvatarElement = styled.div`
    background-color: ${AVATAR_BACKGROUND};
    border-radius: 50%;
    color: ${AVATAR_COLOR};
    display: inline-block;
    font-size: ${AVATAR_FONT_SIZE};
    height: ${AVATAR_SIZE};
    overflow: hidden;
    position: relative;
    text-align: center;
    vertical-align: middle;
    width: ${AVATAR_SIZE};

    & > svg {
        fill: currentColor;
        height: ${AVATAR_SIZE};
        width: 1em;
    }

    & > img {
        height: auto;
        max-width: 100%;
    }
`;

const contentStyle = css`
    display: block;
    line-height: ${AVATAR_SIZE};
    width: 100%;
`;

export const IconElement = styled(FontIcon)`${contentStyle}`;
export const TitleElement = styled.span`${contentStyle}`;
