// mews-ui/lib/Calendar/style.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _templateObject = _taggedTemplateLiteral(['\n    display: inline-flex;\n'], ['\n    display: inline-flex;\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    position: relative;\n    overflow-x: hidden;\n'], ['\n    position: relative;\n    overflow-x: hidden;\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    position: absolute;\n    top: 0;\n    left: 0;\n'], ['\n    position: absolute;\n    top: 0;\n    left: 0;\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n    position: absolute;\n    top: 0;\n    right: 0;\n'], ['\n    position: absolute;\n    top: 0;\n    right: 0;\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n    width: 180px;\n    height: ', 'px;\n    border-right: 1px solid ', ';\n    overflow: hidden;\n'], ['\n    width: 180px;\n    height: ', 'px;\n    border-right: 1px solid ', ';\n    overflow: hidden;\n']),
    _templateObject6 = _taggedTemplateLiteral(['\n    width: 200px;\n    height: 100%;\n    overflow-y: scroll;\n    overflow-x: hidden;\n'], ['\n    width: 200px;\n    height: 100%;\n    overflow-y: scroll;\n    overflow-x: hidden;\n']),
    _templateObject7 = _taggedTemplateLiteral(['\n    font-size: 14px;\n    width: 100%;\n    height: 48px;\n    display: flex;\n    align-items: center;\n    justify-content: center;\n    cursor: pointer;\n\n    ', '\n'], ['\n    font-size: 14px;\n    width: 100%;\n    height: 48px;\n    display: flex;\n    align-items: center;\n    justify-content: center;\n    cursor: pointer;\n\n    ', '\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n    display: flex;\n\n    ', '\n'], ['\n    display: flex;\n\n    ', '\n']),
    _templateObject9 = _taggedTemplateLiteral(['\n    height: ', 'px;\n    width: ', 'px;\n    padding: ', 'px ', 'px;\n\n    box-sizing: border-box;\n\n    #overflow: hidden;\n\n    display: inline-flex;\n    flex-direction: column;\n    align-items: center;\n    justify-content: flex-start;\n'], ['\n    height: ', 'px;\n    width: ', 'px;\n    padding: ', 'px ', 'px;\n\n    box-sizing: border-box;\n\n    #overflow: hidden;\n\n    display: inline-flex;\n    flex-direction: column;\n    align-items: center;\n    justify-content: flex-start;\n']),
    _templateObject10 = _taggedTemplateLiteral(['\n    display: flex;\n    justify-content: space-between;\n'], ['\n    display: flex;\n    justify-content: space-between;\n']),
    _templateObject11 = _taggedTemplateLiteral(['\n    font-size: 14px;\n    font-weight: 500;\n    text-align: center;\n    height: ', 'px;\n    margin-top: -8px;\n    margin-bottom: ', 'px;\n\n    align-items: center;\n    display: flex;\n    justify-content: center;\n'], ['\n    font-size: 14px;\n    font-weight: 500;\n    text-align: center;\n    height: ', 'px;\n    margin-top: -8px;\n    margin-bottom: ', 'px;\n\n    align-items: center;\n    display: flex;\n    justify-content: center;\n']),
    _templateObject12 = _taggedTemplateLiteral(['\n    color: ', ';\n    font-size: 12px;\n    font-weight: 500;\n    position: absolute;\n    top: ', 'px;\n'], ['\n    color: ', ';\n    font-size: 12px;\n    font-weight: 500;\n    position: absolute;\n    top: ', 'px;\n']),
    _templateObject13 = _taggedTemplateLiteral(['\n    padding: 0 12px;\n    width: ', 'px;\n    box-sizing: border-box;\n    display: inline-flex;\n    justify-content: space-between;\n'], ['\n    padding: 0 12px;\n    width: ', 'px;\n    box-sizing: border-box;\n    display: inline-flex;\n    justify-content: space-between;\n']),
    _templateObject14 = _taggedTemplateLiteral(['\n    width: ', 'px;\n    height: ', 'px;\n    text-align: center;\n'], ['\n    width: ', 'px;\n    height: ', 'px;\n    text-align: center;\n']),
    _templateObject15 = _taggedTemplateLiteral(['\n    background: transparent;\n    border: 0;\n    color: ', ';\n    cursor: pointer;\n    display: inline-block;\n    font-family: ', ';\n    font-size: 12px;\n    outline: none;\n    -webkit-tap-highlight-color: transparent;\n\n    width: 40px;\n    height: 40px;\n    margin: 0;\n\n    ', '\n\n    &[disabled] {\n        color: ', ';\n        cursor: auto;\n        pointer-events: none;\n    }\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n'], ['\n    background: transparent;\n    border: 0;\n    color: ', ';\n    cursor: pointer;\n    display: inline-block;\n    font-family: ', ';\n    font-size: 12px;\n    outline: none;\n    -webkit-tap-highlight-color: transparent;\n\n    width: 40px;\n    height: 40px;\n    margin: 0;\n\n    ', '\n\n    &[disabled] {\n        color: ', ';\n        cursor: auto;\n        pointer-events: none;\n    }\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n']),
    _templateObject16 = _taggedTemplateLiteral(['\n        &:hover {\n            background-color: ', ';\n            color: ', ' !important;\n        }\n    '], ['\n        &:hover {\n            background-color: ', ';\n            color: ', ' !important;\n        }\n    ']),
    _templateObject17 = _taggedTemplateLiteral(['\n        border-radius: 50%;\n        background-color: ', ';\n        color: ', ';\n    '], ['\n        border-radius: 50%;\n        background-color: ', ';\n        color: ', ';\n    ']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled, { css } from 'styled-components';
import { COLOR_WHITE, PALETTE_GREY_400 } from '../colors';
import { DEFAULT_THEME, PREFERRED_FONT } from '../variables';
import { alpha } from '../utils/color';
import IconButton from '../Button/IconButton';

var MONTH_NAME_SIZE = 48;
var DAY_SIZE = 40;
var DAY_NAME_SIZE = 20;
var HORIZONTAL_PADDING = 12;
var VERTICAL_PADDING = 8;

export var MONTH_WIDTH = 7 * DAY_SIZE + HORIZONTAL_PADDING * 2;
var MONTH_HEIGHT = MONTH_NAME_SIZE + DAY_NAME_SIZE + 6 * DAY_SIZE + VERTICAL_PADDING * 2;

export var CalendarElement = styled.div(_templateObject);

export var CalendarMonthGridContainerElement = styled.div(_templateObject2);

var buttonDefaultProps = _extends({}, IconButton.defaultProps, {
    tabIndex: -1
});

export var PrevMonthIconButton = styled(IconButton)(_templateObject3);
PrevMonthIconButton.defaultProps = _extends({}, buttonDefaultProps, { icon: 'chevron_left' });

export var NextMonthIconButton = styled(IconButton)(_templateObject4);
NextMonthIconButton.defaultProps = _extends({}, buttonDefaultProps, { icon: 'chevron_right' });

export var CalendarMonthSelectorElement = styled.div(_templateObject5, MONTH_HEIGHT, function (props) {
    return props.theme.dividerColor;
});
CalendarMonthSelectorElement.defaultProps = { theme: DEFAULT_THEME };

export var ScrollContainerElement = styled.div(_templateObject6);

export var SelectableMonthElement = styled.div.attrs({
    onClick: function onClick(props) {
        return function () {
            props.onMonthClick(props.month);
        };
    }
})(_templateObject7, function (props) {
    return props.selected && '\n        font-weight: 500;\n        color: ' + COLOR_WHITE + ';\n        background-color: ' + props.theme.primaryColor + ';\n    ';
});
SelectableMonthElement.defaultProps = { theme: DEFAULT_THEME };

export var MonthGridElement = styled.div(_templateObject8, function (props) {
    return props.animating && '\n        transition: transform .2s;\n    ';
});

export var MonthElement = styled.div(_templateObject9, MONTH_HEIGHT, MONTH_WIDTH, VERTICAL_PADDING, HORIZONTAL_PADDING);

export var WeekElement = styled.div(_templateObject10);

export var MonthNameElement = styled.div(_templateObject11, MONTH_NAME_SIZE, DAY_NAME_SIZE);

export var DayNamesElement = styled.div(_templateObject12, PALETTE_GREY_400, MONTH_NAME_SIZE);

export var DayNamesWeekElement = styled.div(_templateObject13, MONTH_WIDTH);

export var DayNameElement = styled.span(_templateObject14, DAY_SIZE, DAY_NAME_SIZE);

export var CalendarDayElement = styled.div(_templateObject14, DAY_SIZE, DAY_SIZE);

export var DayButtonElement = styled.button(_templateObject15, function (props) {
    return props.theme.primaryTextColor;
}, PREFERRED_FONT, function (props) {
    return props.singlePicker && '\n        width: 32px;\n        height: 32px;\n        margin: 4px;\n        padding: 0;\n        border-radius: 50%;\n    ';
}, function (props) {
    return props.theme.disabledColor;
}, function (props) {
    return !props.selected && css(_templateObject16, alpha(props.theme.primaryColor, 0.3), COLOR_WHITE);
}, function (props) {
    return props.selected && css(_templateObject17, props.theme.primaryColor, COLOR_WHITE);
}, function (props) {
    return props.startDate && '\n        background-color: ' + props.theme.primaryColor + ';\n        color: ' + COLOR_WHITE + ';\n        border-top-left-radius: 50%;\n        border-bottom-left-radius: 50%;\n    ';
}, function (props) {
    return props.endDate && '\n        background-color: ' + props.theme.primaryColor + ';\n        color: ' + COLOR_WHITE + ';\n        border-top-right-radius: 50%;\n        border-bottom-right-radius: 50%;\n    ';
}, function (props) {
    return props.inBetweenDate && '\n        background-color: ' + alpha(props.theme.primaryColor, 0.6) + ';\n        color: ' + COLOR_WHITE + ';\n    ';
});
DayButtonElement.defaultProps = {
    theme: DEFAULT_THEME
};
