// mews-ui/src/Menu/style.jsx
import styled from 'styled-components';
import { ANIMATION_CURVE_DEFAULT, ANIMATION_DURATION, DEFAULT_THEME, FONT_SIZE_LARGE, FONT_SIZE_NORMAL } from '../variables';

const DROP_DOWN_MENU_ARROW_SIZE = '5px';

export const DropDownMenuElement = styled.div`
    align-items: center;
    cursor: pointer;
    display: flex;
    font-size: ${props => (props.large ? FONT_SIZE_LARGE : FONT_SIZE_NORMAL)};
    min-height: ${props => (props.large ? '30px' : '19px')};
    outline: none;
    position: relative;
`;

export const DropDownMenuValueElement = styled.span`
    color: ${props => props.theme.primaryTextColor};
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
`;
DropDownMenuValueElement.defaultProps = { theme: DEFAULT_THEME };

export const DropDownMenuArrowElement = styled.span`
    border-left: ${DROP_DOWN_MENU_ARROW_SIZE} solid transparent;
    border-right: ${DROP_DOWN_MENU_ARROW_SIZE} solid transparent;
    border-top: ${DROP_DOWN_MENU_ARROW_SIZE} solid ${props => props.theme.hintTextColor};
    height: 0;
    margin-left: auto;
    pointer-events: none;
    transition: transform ${ANIMATION_DURATION} ${ANIMATION_CURVE_DEFAULT};
    width: 0;
`;
DropDownMenuArrowElement.defaultProps = { theme: DEFAULT_THEME };

export const IconMenuElement = styled.div`
    display: inline-block;
    position: relative;
`;
