// mews-ui/src/TimePicker/style.jsx
import styled from 'styled-components';
import { ANIMATION_CURVE_DEFAULT, DEFAULT_THEME } from '../variables';
import Slider from '../Slider/Slider';
import RadioGroup from '../Radio/RadioGroup';
import RadioButton from '../Radio/RadioButton';
import { InputElement as x } from '../Input/style';

export const ITEM_HEIGHT = 48;

export const TimePickerElement = styled.div`
    padding: 8px;
    box-sizing: border-box;

    display: flex;
`;

export const MeridiemSelectorElement = styled(RadioGroup)`
    margin-left: calc(25% - 16px);
`;

export const MeridiemButtonElement = RadioButton;

export const NumberSelectorElement = styled.div`
    display: flex;
    align-items: center;
`;

export const LabelElement = styled.div`
    min-width: 25%;
`;

export const SliderElement = Slider;
SliderElement.defaultProps = {
    ...SliderElement.defaultProps,
    step: 1,
};

export const InputElement = styled(x).attrs({
    type: 'number',
    step: 1,
})`
    width: 48px;
    margin-left: 16px;
    border-bottom: 1px solid ${props => props.theme.hintTextColor};
    padding-bottom: 4px;

    &:focus {
        border-color: ${props => props.theme.primaryColor};
    }
`;

InputElement.defaultProps = {
    ...InputElement.defaultProps,
    theme: DEFAULT_THEME,
};

export const ScrollableSelectorElement = styled.div`
    width: 64px;
    height: ${ITEM_HEIGHT * 5}px;
    overflow: hidden;
    position: relative;
    touch-action: none;
`;

export const ScrollContainerElement = styled.div`
    width: 100%;
    height: 100%;
    position: absolute;
    left: 0;
    top: ${ITEM_HEIGHT * 2}px;
    transition: transform 200ms ${ANIMATION_CURVE_DEFAULT};
`;

export const ScrollItemHighlightElement = styled.div`
    position: absolute;
    left: 0;
    top: ${ITEM_HEIGHT * 2}px;
    background-color: ${props => props.theme.primaryColor};
    height: ${ITEM_HEIGHT}px;
    width: 100%;
    box-sizing: border-box;
    pointer-events: none;
    display: flex;
    align-items: center;
    justify-content: flex-end;
    font-weight: 500;
    color: #fff;
`;
ScrollItemHighlightElement.defaultProps = { theme: DEFAULT_THEME };

export const ScrollItemElement = styled.div.attrs({
    onClick: props => () => props.onItemClick(props.value),
})`
    font-size: 14px;
    width: 100%;
    height: ${ITEM_HEIGHT}px;
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
    box-sizing: border-box;
    user-select: none;

    ${props => props.selected && `
        font-weight: 500;
        color: #fff;
    `}
`;
