// mews-ui/src/TransitionGroup/StyledTransitionGroup.js
import React, { Component } from 'react';
import { PropTypes as pt } from 'prop-types';
import ReactDOM from 'react-dom';
import { TransitionGroup } from 'react-transition-group';
import StyledTransitionGroupChild from './StyledTransitionGroupChild';

export default class StyledTransitionGroup extends Component {
    static propTypes = {
        children: pt.node,
        childrenProps: pt.object, // eslint-disable-line react/forbid-prop-types
        transitionAppear: pt.bool,
        transitionEnter: pt.bool,
        transitionLeave: pt.bool,
    };

    static defaultProps = {
        transitionAppear: false,
        transitionEnter: true,
        transitionLeave: true,
    };

    wrapChild = child => {
        const { childrenProps, transitionAppear, transitionEnter, transitionLeave } = this.props;
        const props = {
            children: child,
            childrenProps,
            transitionAppear,
            transitionEnter,
            transitionLeave,
        };

        return (
            <StyledTransitionGroupChild {...props} />
        );
    };

    getDomElement() {
        return ReactDOM.findDOMNode(this);
    }

    render() {
        const props = {
            ...this.props,
            childFactory: this.wrapChild,
        };

        delete props.innerRef;
        delete props.childrenProps;

        return (
            <TransitionGroup {...props}>{this.props.children}</TransitionGroup>
        );
    }
}
