// mews-ui/lib/List/style.js
var _templateObject = _taggedTemplateLiteral(['\n    outline: none;\n    padding: ', ' 0;\n    position: relative;\n    text-align: left;\n'], ['\n    outline: none;\n    padding: ', ' 0;\n    position: relative;\n    text-align: left;\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    outline: none;\n'], ['\n    outline: none;\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    align-items: center;\n    box-sizing: border-box;\n    display: flex;\n    outline: none;\n    padding: ', ' ', ';\n    position: relative;\n    height: ', ';\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n\n    &:focus {\n        background-color: ', ';\n    }\n\n    &:hover {\n        background-color: ', ';\n        cursor: pointer;\n    }\n'], ['\n    align-items: center;\n    box-sizing: border-box;\n    display: flex;\n    outline: none;\n    padding: ', ' ', ';\n    position: relative;\n    height: ', ';\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n\n    ', '\n\n    &:focus {\n        background-color: ', ';\n    }\n\n    &:hover {\n        background-color: ', ';\n        cursor: pointer;\n    }\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n        height: ', ';\n    '], ['\n        height: ', ';\n    ']),
    _templateObject5 = _taggedTemplateLiteral(['\n        color: ', ';\n    '], ['\n        color: ', ';\n    ']),
    _templateObject6 = _taggedTemplateLiteral(['\n        display: none;\n    '], ['\n        display: none;\n    ']),
    _templateObject7 = _taggedTemplateLiteral(['\n    align-items: center;\n    display: flex;\n'], ['\n    align-items: center;\n    display: flex;\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n    align-items: center;\n    display: flex;\n    margin-left: auto;\n    padding-left: ', ';\n'], ['\n    align-items: center;\n    display: flex;\n    margin-left: auto;\n    padding-left: ', ';\n']),
    _templateObject9 = _taggedTemplateLiteral(['\n    color: ', ';\n    \n    ', '\n    ', '\n'], ['\n    color: ', ';\n    \n    ', '\n    ', '\n']),
    _templateObject10 = _taggedTemplateLiteral(['\n    margin-right: 16px;\n'], ['\n    margin-right: 16px;\n']),
    _templateObject11 = _taggedTemplateLiteral(['\n    display: flex;\n    flex-flow: column;\n    overflow-x: hidden;\n'], ['\n    display: flex;\n    flex-flow: column;\n    overflow-x: hidden;\n']),
    _templateObject12 = _taggedTemplateLiteral(['\n    font-size: ', ';\n    overflow-x: hidden;\n    text-overflow: ellipsis;\n    white-space: nowrap;\n'], ['\n    font-size: ', ';\n    overflow-x: hidden;\n    text-overflow: ellipsis;\n    white-space: nowrap;\n']),
    _templateObject13 = _taggedTemplateLiteral(['\n    color: ', ';\n    font-size: ', ';\n    overflow-x: hidden;\n    text-overflow: ellipsis;\n    word-wrap: nowrap;\n'], ['\n    color: ', ';\n    font-size: ', ';\n    overflow-x: hidden;\n    text-overflow: ellipsis;\n    word-wrap: nowrap;\n']),
    _templateObject14 = _taggedTemplateLiteral(['\n    position: relative;\n    border-bottom: 1px solid ', ';\n    padding: 8px 16px;\n'], ['\n    position: relative;\n    border-bottom: 1px solid ', ';\n    padding: 8px 16px;\n']),
    _templateObject15 = _taggedTemplateLiteral(['\n    position: absolute;\n    top: 8px;\n    color: ', ';\n    pointer-events: none;\n'], ['\n    position: absolute;\n    top: 8px;\n    color: ', ';\n    pointer-events: none;\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled, { css } from 'styled-components';
import { PALETTE_GREY_200 } from '../colors';
import { COLOR_TEXT_SECONDARY, FONT_SIZE_SMALL, FONT_SIZE_NORMAL, DEFAULT_THEME } from '../variables';
import { InputElement } from '../Input/style';

var LIST_VERTICAL_PADDING = function LIST_VERTICAL_PADDING(props) {
    return props.dense ? '4px' : '8px';
};
var LIST_HORIZONTAL_PADDING = '16px';
var LIST_ITEM_HOVER_COLOR = PALETTE_GREY_200;
var LIST_ITEM_ACTIVE_BACKGROUND = PALETTE_GREY_200;

export var ListElement = styled.div(_templateObject, LIST_VERTICAL_PADDING);

export var ListItemsWrapperElement = styled.div(_templateObject2);

export var ListItemElement = styled.div(_templateObject3, LIST_VERTICAL_PADDING, LIST_HORIZONTAL_PADDING, function (props) {
    return props.dense ? '40px' : '48px';
}, function (props) {
    return props.singleLineWithAvatar && css(_templateObject4, props.dense ? '48px' : '56px');
}, function (props) {
    return props.twoLine && css(_templateObject4, props.dense ? '60px' : '72px');
}, function (props) {
    return props.threeLine && css(_templateObject4, props.dense ? '76px' : '88px');
}, function (props) {
    return props.selected && css(_templateObject5, props.theme.primaryColor);
}, function (props) {
    return props.hidden && css(_templateObject6);
}, LIST_ITEM_ACTIVE_BACKGROUND, LIST_ITEM_HOVER_COLOR);
ListItemElement.defaultProps = {
    theme: DEFAULT_THEME
};

export var ListItemWrapper = styled.span(_templateObject2);

export var ListItemLeftElement = styled.span(_templateObject7);

export var ListItemRightElement = styled.span(_templateObject8, LIST_HORIZONTAL_PADDING);

export var ListItemIconElement = styled.span(_templateObject9, COLOR_TEXT_SECONDARY, function (props) {
    return props.left && 'margin-right: 32px;';
}, function (props) {
    return props.right && 'margin-left: 16px;';
});

export var ListItemAvatarElement = styled.span(_templateObject10);

export var ListItemContentElement = styled.span(_templateObject11);

export var ListItemPrimaryTextElement = styled.span(_templateObject12, function (props) {
    return props.dense ? FONT_SIZE_SMALL : FONT_SIZE_NORMAL;
});

export var ListItemSecondaryTextElement = styled.span(_templateObject13, COLOR_TEXT_SECONDARY, FONT_SIZE_SMALL);

export var FilterElement = styled.div(_templateObject14, function (props) {
    return props.theme.dividerColor;
});
FilterElement.defaultProps = { theme: DEFAULT_THEME };
export var FilterInputElement = InputElement;
export var FilterPlaceholderElement = styled.div(_templateObject15, function (props) {
    return props.theme.hintTextColor;
});
FilterPlaceholderElement.defaultProps = { theme: DEFAULT_THEME };
