// mews-ui/lib/ProgressBar/ProgressCircular.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import { PropTypes as pt } from 'prop-types';
import ProgressMode from './ProgressMode';
import { calculateRatio, ProgressCircularElement, ProgressCircularPath, ProgressCircularSvg } from './style';

var ProgressCircular = function (_Component) {
    _inherits(ProgressCircular, _Component);

    function ProgressCircular() {
        _classCallCheck(this, ProgressCircular);

        return _possibleConstructorReturn(this, (ProgressCircular.__proto__ || Object.getPrototypeOf(ProgressCircular)).apply(this, arguments));
    }

    _createClass(ProgressCircular, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                max = _props.max,
                min = _props.min,
                mode = _props.mode,
                value = _props.value;

            var svgProps = {
                mode: mode,
                viewBox: '0 0 60 60'
            };
            var pathProps = {
                cx: 30,
                cy: 30,
                r: 25,
                mode: mode
            };

            if (mode === ProgressMode.Determinate) {
                pathProps.style = {
                    strokeDasharray: 2 * Math.PI * 25 * calculateRatio({ max: max, min: min, value: value }) + ', 400'
                };
            }

            return React.createElement(
                ProgressCircularElement,
                { 'aria-valuenow': value, 'aria-valuemin': min, 'aria-valuemax': max },
                React.createElement(
                    ProgressCircularSvg,
                    svgProps,
                    React.createElement(ProgressCircularPath, pathProps)
                )
            );
        }
    }]);

    return ProgressCircular;
}(Component);

ProgressCircular.propTypes = {
    max: pt.number,
    min: pt.number,
    mode: pt.instanceOf(ProgressMode),
    value: pt.number
};
ProgressCircular.defaultProps = {
    max: 100,
    min: 0,
    mode: ProgressMode.Indeterminate,
    value: 0
};
export default ProgressCircular;
