// mews-ui/lib/ProgressBar/style.js
var _templateObject = _taggedTemplateLiteral(['\n    0% {\n        transform: translate(-50%) scaleX(0);\n    }\n\n    50% {\n        transform: translate(-0%) scaleX(0.3);\n    }\n\n    100% {\n        transform: translate(50%) scaleX(0);\n    }\n'], ['\n    0% {\n        transform: translate(-50%) scaleX(0);\n    }\n\n    50% {\n        transform: translate(-0%) scaleX(0.3);\n    }\n\n    100% {\n        transform: translate(50%) scaleX(0);\n    }\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    background: ', ';\n    display: inline-block;\n    height: ', ';\n    overflow: hidden;\n    position: relative;\n    width: 100%;\n'], ['\n    background: ', ';\n    display: inline-block;\n    height: ', ';\n    overflow: hidden;\n    position: relative;\n    width: 100%;\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    bottom: 0;\n    left: 0;\n    position: absolute;\n    right: 0;\n    top: 0;\n    transform: scaleX(0);\n    transform-origin: left center;\n    transition-duration: ', ';\n    transition-timing-function: ', ';\n    background-color: ', ';\n\n    ', '\n\n    ', '\n'], ['\n    bottom: 0;\n    left: 0;\n    position: absolute;\n    right: 0;\n    top: 0;\n    transform: scaleX(0);\n    transform-origin: left center;\n    transition-duration: ', ';\n    transition-timing-function: ', ';\n    background-color: ', ';\n\n    ', '\n\n    ', '\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n        animation: ', ' 1s linear infinite;\n        transform-origin: center center;\n    '], ['\n        animation: ', ' 1s linear infinite;\n        transform-origin: center center;\n    ']),
    _templateObject5 = _taggedTemplateLiteral(['\n        background-color: ', ';\n    '], ['\n        background-color: ', ';\n    ']),
    _templateObject6 = _taggedTemplateLiteral(['\n    100% {\n        transform: rotate(360deg);\n    }\n'], ['\n    100% {\n        transform: rotate(360deg);\n    }\n']),
    _templateObject7 = _taggedTemplateLiteral(['\n    0% {\n        stroke-dasharray: ', ', ', ';\n        stroke-dashoffset: ', ';\n    }\n\n    50% {\n        stroke-dasharray: ', ' * 89), ', ';\n        stroke-dashoffset: ', ' * -35);\n    }\n\n    100% {\n        stroke-dasharray: ', ', ', ';\n        stroke-dashoffset: ', ';\n    }\n'], ['\n    0% {\n        stroke-dasharray: ', ', ', ';\n        stroke-dashoffset: ', ';\n    }\n\n    50% {\n        stroke-dasharray: ', ' * 89), ', ';\n        stroke-dashoffset: ', ' * -35);\n    }\n\n    100% {\n        stroke-dasharray: ', ', ', ';\n        stroke-dashoffset: ', ';\n    }\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n    display: inline-block;\n    height: ', ';\n    position: relative;\n    transform: rotate(-90deg);\n    width: ', ';\n'], ['\n    display: inline-block;\n    height: ', ';\n    position: relative;\n    transform: rotate(-90deg);\n    width: ', ';\n']),
    _templateObject9 = _taggedTemplateLiteral(['\n    height: 100%;\n    width: 100%;\n\n    ', '\n'], ['\n    height: 100%;\n    width: 100%;\n\n    ', '\n']),
    _templateObject10 = _taggedTemplateLiteral(['\n        animation: ', ' 2s linear infinite;\n    '], ['\n        animation: ', ' 2s linear infinite;\n    ']),
    _templateObject11 = _taggedTemplateLiteral(['\n    fill: none;\n    stroke: ', ';\n    stroke-dasharray: 0, calc(', ' * 200);\n    stroke-dashoffset: 0;\n    stroke-linecap: round;\n    stroke-miterlimit: 20;\n    stroke-width: 4;\n    transition: stroke-dasharray ', ' ', ';\n\n    ', '\n'], ['\n    fill: none;\n    stroke: ', ';\n    stroke-dasharray: 0, calc(', ' * 200);\n    stroke-dashoffset: 0;\n    stroke-linecap: round;\n    stroke-miterlimit: 20;\n    stroke-width: 4;\n    transition: stroke-dasharray ', ' ', ';\n\n    ', '\n']),
    _templateObject12 = _taggedTemplateLiteral(['\n        animation: ', ' 1.5s ease-in-out infinite;\n        stroke-dasharray: calc(', ' * 1), calc(', ' * 200);\n        stroke-dashoffset: 0;\n    '], ['\n        animation: ', ' 1.5s ease-in-out infinite;\n        stroke-dasharray: calc(', ' * 1), calc(', ' * 200);\n        stroke-dashoffset: 0;\n    ']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled, { css, keyframes } from 'styled-components';

import { PALETTE_GREY_500 } from '../colors';
import { ANIMATION_CURVE_DEFAULT, ANIMATION_DURATION, DEFAULT_THEME } from '../variables';
import ProgressMode from './ProgressMode';

var PROGRESS_HEIGHT = '4px';
export var PROGRESS_BAR_COLOR = PALETTE_GREY_500;

var CIRCLE_WRAPPER_WIDTH = '60px';
var CIRCLE_RADIUS = 25;
var SCALE_RATIO = CIRCLE_RADIUS / 20;

var linearIndeterminateBar = keyframes(_templateObject);

export var ProgressLinearElement = styled.div(_templateObject2, PROGRESS_BAR_COLOR, PROGRESS_HEIGHT);

export var LinearValueBar = styled.span(_templateObject3, ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.theme.primaryColor;
}, function (props) {
    return props.mode === ProgressMode.Indeterminate && css(_templateObject4, linearIndeterminateBar);
}, function (props) {
    return props.disabled && css(_templateObject5, PROGRESS_BAR_COLOR);
});
LinearValueBar.defaultProps = { theme: DEFAULT_THEME };

var circularIndeterminateBarRotate = keyframes(_templateObject6);

var circularIndeterminateBarDash = keyframes(_templateObject7, SCALE_RATIO * 1, SCALE_RATIO * 200, SCALE_RATIO * 0, SCALE_RATIO, SCALE_RATIO * 200, SCALE_RATIO, SCALE_RATIO * 89, SCALE_RATIO * 200, SCALE_RATIO * -124);

export var ProgressCircularElement = styled.div(_templateObject8, CIRCLE_WRAPPER_WIDTH, CIRCLE_WRAPPER_WIDTH);

export var ProgressCircularSvg = styled.svg(_templateObject9, function (props) {
    return props.mode === ProgressMode.Indeterminate && css(_templateObject10, circularIndeterminateBarRotate);
});

export var ProgressCircularPath = styled.circle(_templateObject11, function (props) {
    return props.theme.primaryColor;
}, SCALE_RATIO, ANIMATION_DURATION, ANIMATION_CURVE_DEFAULT, function (props) {
    return props.mode === ProgressMode.Indeterminate && css(_templateObject12, circularIndeterminateBarDash, SCALE_RATIO, SCALE_RATIO);
});
ProgressCircularPath.defaultProps = { theme: DEFAULT_THEME };

export function calculateRatio(_ref) {
    var max = _ref.max,
        min = _ref.min,
        value = _ref.value;

    if (value < min) {
        return 0;
    }
    if (value > max) {
        return 1;
    }

    return (value - min) / (max - min);
}
