// mews-ui/src/OpenableOverlay/Origin.js
import { enumify } from 'mews-js';

export default function Origin() {}
enumify(Origin, {
    TopLeft: {
        value: 'top left',
        vertical: 'top',
        horizontal: 'left',
    },
    TopCenter: {
        value: 'top center',
        vertical: 'top',
        horizontal: 'center',
    },
    TopRight: {
        value: 'top right',
        vertical: 'top',
        horizontal: 'right',
    },
    CenterLeft: {
        value: 'center left',
        vertical: 'center',
        horizontal: 'left',
    },
    CenterCenter: {
        value: 'center center',
        vertical: 'center',
        horizontal: 'center',
    },
    CenterRight: {
        value: 'center right',
        vertical: 'center',
        horizontal: 'right',
    },
    BottomLeft: {
        value: 'bottom left',
        vertical: 'bottom',
        horizontal: 'left',
    },
    BottomCenter: {
        value: 'bottom center',
        vertical: 'bottom',
        horizontal: 'center',
    },
    BottomRight: {
        value: 'bottom right',
        vertical: 'bottom',
        horizontal: 'right',
    },
});
