// mews-ui/src/OpenableOverlay/style.jsx
import styled from 'styled-components';
import { COLOR_WHITE } from '../colors';
import { ANIMATION_CURVE_DEFAULT, ANIMATION_DURATION, Z_INDEX_HIGH, Z_DEPTH_SHADOW_1 } from '../variables';

const OPENABLE_OVERLAY_BACKGROUND = COLOR_WHITE;

export const OpenableOverlayElement = styled.div`
    background-color: ${OPENABLE_OVERLAY_BACKGROUND};
    box-shadow: ${Z_DEPTH_SHADOW_1};
    left: ${props => props.placement.left}px;
    max-height: 0;
    opacity: 0;
    overflow-x: hidden;
    overflow-y: auto;
    outline: none;
    position: absolute;
    top: ${props => props.placement.top}px;
    transform: scale(0);
    transform-origin: ${props => props.origin};
    transition:
        visibility 0s linear ${ANIMATION_DURATION},
        opacity ${ANIMATION_DURATION} ${ANIMATION_CURVE_DEFAULT},
        transform ${ANIMATION_DURATION} ${ANIMATION_CURVE_DEFAULT};
    visibility: hidden;
    z-index: ${Z_INDEX_HIGH};

    max-height: ${props => props.maxHeight}px;

    ${props => props.opened && `
        visibility: visible;
        opacity: 1;
        transform: scale(1);
        transition-delay: 0s;
    `}

`;

/*
    transition:
        visibility 0s linear ${ANIMATION_DURATION},
        opacity ${ANIMATION_DURATION} ${ANIMATION_CURVE_DEFAULT},
        transform ${ANIMATION_DURATION} ${ANIMATION_CURVE_DEFAULT};
        */
