// mews-ui/src/Ripple/style.jsx
import styled, { css } from 'styled-components';

import { Z_INDEX_NORMAL, Z_INDEX_HIGH } from '../variables';

const RIPPLE_DURATION = '800ms';
const RIPPLE_FINAL_OPACITY = '0.3';

export const RIPPLE_WRAPPER_CLASSNAME = 'rippleWrapper';

export const RippleWrapper = styled.span`
    bottom: 0;
    left: 0;
    pointer-events: none;
    position: absolute;
    right: 0;
    top: 0;
    z-index: ${Z_INDEX_NORMAL}
`;

export const RippleElement = styled.span`
    background-color: currentColor;
    border-radius: 50%;
    height: ${({ width }) => width + 'px'};
    left: 50%;
    pointer-events: none;
    position: absolute;
    top: 50%;
    transform-origin: 50% 50%;
    transition-duration: ${RIPPLE_DURATION};
    width: ${({ width }) => width + 'px'};
    z-index: ${Z_INDEX_HIGH};

    ${({ width, left, top, restarting, scale = restarting ? 0 : 1 }) => css`
        transform: translate3d(${-(width / 2) + left}px, ${-(width / 2) + top}px, 0) scale(${scale});
    `}

    ${({ restarting }) => restarting && css`
        opacity: ${RIPPLE_FINAL_OPACITY};
        transition-property: none;
    `}

    ${({ active }) => active && css`
        opacity: ${RIPPLE_FINAL_OPACITY};
        transition-property: transform;
    `}

    ${({ active, restarting }) => !(active || restarting) && `
        opacity: 0;
        transition-property: opacity, transform;
    `}
`;
