// mews-ui/lib/Slider/style.js
var _templateObject = _taggedTemplateLiteral(['\n    cursor: pointer;\n    height: ', ';\n    outline: none;\n    position: relative;\n    touch-action: none;\n    user-select: none;\n    width: 100%;\n'], ['\n    cursor: pointer;\n    height: ', ';\n    outline: none;\n    position: relative;\n    touch-action: none;\n    user-select: none;\n    width: 100%;\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n    position: absolute;\n    left: 0;\n    top: 50%;\n    height: 2px;\n\n    ', ' {\n        transition-duration: 0s;\n    }\n'], ['\n    position: absolute;\n    left: 0;\n    top: 50%;\n    height: 2px;\n\n    ', ' {\n        transition-duration: 0s;\n    }\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n    align-items: center;\n    background-color: transparent;\n    display: inline-flex;\n    height: ', ';\n    justify-content: center;\n    width: 20px;\n'], ['\n    align-items: center;\n    background-color: transparent;\n    display: inline-flex;\n    height: ', ';\n    justify-content: center;\n    width: 20px;\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n    fill: ', ';\n    stroke: ', ';\n    stroke-width: 3.5;\n    position: absolute;\n    transform: scale(0.571);\n    transition: transform 0.1s ease-out, fill 0.1s ease-out, stroke 0.1s ease-out;\n\n    ', '\n\n    ', '\n'], ['\n    fill: ', ';\n    stroke: ', ';\n    stroke-width: 3.5;\n    position: absolute;\n    transform: scale(0.571);\n    transition: transform 0.1s ease-out, fill 0.1s ease-out, stroke 0.1s ease-out;\n\n    ', '\n\n    ', '\n']),
    _templateObject5 = _taggedTemplateLiteral([''], ['']),
    _templateObject6 = _taggedTemplateLiteral(['\n    background-color: ', ';\n    border-radius: 50%;\n    height: 21px;\n    opacity: 0;\n    width: 21px;\n\n    transition: transform 266.67ms ease-out, opacity 266.67ms ease-out, background-color 266.67ms ease-out;\n    \n    ', '\n'], ['\n    background-color: ', ';\n    border-radius: 50%;\n    height: 21px;\n    opacity: 0;\n    width: 21px;\n\n    transition: transform 266.67ms ease-out, opacity 266.67ms ease-out, background-color 266.67ms ease-out;\n    \n    ', '\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import React from 'react';
import styled from 'styled-components';
import { COLOR_WHITE } from '../colors';
import { DEFAULT_THEME } from '../variables';
import ProgressLinear from '../ProgressBar/ProgressLinear';
import ProgressMode from '../ProgressBar/ProgressMode';
import { PROGRESS_BAR_COLOR, LinearValueBar } from '../ProgressBar/style';

var SLIDER_HEIGHT = '48px';

export var SliderElement = styled.div(_templateObject, SLIDER_HEIGHT);

export var SliderTrackElement = styled(ProgressLinear).attrs({
    mode: ProgressMode.Determinate
})(_templateObject2, LinearValueBar);

export var KnobContainerElement = styled.div(_templateObject3, SLIDER_HEIGHT);

var KnobSvgElement = styled.svg.attrs({
    width: 21,
    height: 21
})(_templateObject4, function (props) {
    return props.theme.primaryColor;
}, function (props) {
    return props.theme.primaryColor;
}, function (props) {
    return props.active && '\n        transform: scale3d(1, 1, 1);\n    ';
}, function (props) {
    return props.disabled && '\n        stroke: ' + COLOR_WHITE + ';\n        fill: ' + PROGRESS_BAR_COLOR + ';\n    ';
});
KnobSvgElement.defaultProps = { theme: DEFAULT_THEME };

var KnobCircleElement = styled.circle.attrs({
    cx: 10.5,
    cy: 10.5,
    r: 7.875
})(_templateObject5);

export var KnobElement = function KnobElement(props) {
    return React.createElement(
        KnobSvgElement,
        props,
        React.createElement(KnobCircleElement, null)
    );
};

export var KnobFocusElement = styled.div(_templateObject6, function (props) {
    return props.theme.primaryColor;
}, function (props) {
    return props.focused && '\n        transform: scale3d(1.55, 1.55, 1.55);\n        opacity: .25;\n    ';
});
KnobFocusElement.defaultProps = { theme: DEFAULT_THEME };
