// mews-js/src/components/globalization/currencies/currency.js
import _ from 'lodash';
import { type } from '../../../utils';
import { CurrencySymbolPosition, CurrencySymbolSpacing } from '../../../enums';

export default function Currency(data) {
    const symbolPosition = CurrencySymbolPosition.getByName(data.symbolPosition);
    const symbolSpacing = CurrencySymbolSpacing.getByName(data.symbolSpacing);

    const cultureFormats = _.transform(data.cultureFormats, (result, format, cultureName) => {
        result[cultureName] = {
            symbol: format.symbol,
            symbolPosition: CurrencySymbolPosition.getByName(format.symbolPosition),
            symbolSpacing: CurrencySymbolSpacing.getByName(format.symbolSpacing),
        };
    });

    return type.create(Currency, {
        getCode() {
            return data.code;
        },

        getPrecision() {
            return data.precision;
        },

        getSymbol(cultureCode = null) {
            return _.get(cultureFormats, [cultureCode, 'symbol'], data.symbol);
        },

        getSymbolPosition(cultureCode = null) {
            return _.get(cultureFormats, [cultureCode, 'symbolPosition'], symbolPosition);
        },

        getSymbolSpacing(cultureCode = null) {
            return _.get(cultureFormats, [cultureCode, 'symbolSpacing'], symbolSpacing);
        },

        getCultureFormats() {
            return cultureFormats;
        }
    });
}
