// mews-ui/src/Table/style.jsx
import styled, { css } from 'styled-components';
import { DEFAULT_THEME } from '../variables';
import FontIcon from '../FontIcon/FontIcon';

export const TableElement = styled.table`
    border-spacing: 0;
    width: 100%;
`;

export const ColumnDefinitionsElement = styled.colgroup``;

export const ColumnElement = styled.col`
    ${props => props.columnWidth && `
        width: ${props.columnWidth};
    `}
`;

export const TableHeadElement = styled.thead`
    font-size: 12px;
    font-weight: 500;
    color: ${props => props.theme.primaryTextColor};
`;
TableHeadElement.defaultProps = { theme: DEFAULT_THEME };

export const TableBodyElement = styled.tbody`
    font-size: 13px;
    color: ${props => props.theme.primaryTextColor};
`;
TableBodyElement.defaultProps = { theme: DEFAULT_THEME };

export const TableHeaderRowElement = styled.tr`
    padding: 24px 0;
`;

export const TableHeaderCellElement = styled.th`
    border-bottom: 1px solid ${props => props.theme.dividerColor};
    box-sizing: border-box;
    height: 56px;
    padding: 0;
    text-align: ${props => props.columnAlignment || 'left'};

    ${props => props.sortable && `
        cursor: pointer;
    `}
`;
TableHeaderCellElement.defaultProps = { theme: DEFAULT_THEME };

export const TableRowElement = styled.tr``;

export const TableCellElement = styled.td`
    border-bottom: 1px solid ${props => props.theme.dividerColor};
    box-sizing: border-box;
    height: 48px;
    padding: 0;
    text-align: ${props => props.columnAlignment || 'left'};
`;
TableCellElement.defaultProps = { theme: DEFAULT_THEME };

const iconStyle = css`
    color: ${props => props.theme.primaryTextColor}
    font-size: 16px;
    vertical-align: text-bottom;
`;

export const SortAscendingIcon = styled(FontIcon)`${iconStyle}`;
SortAscendingIcon.defaultProps = { value: 'arrow_upward', theme: DEFAULT_THEME };

export const SortDescendingIcon = styled(FontIcon)`${iconStyle}`;
SortDescendingIcon.defaultProps = { value: 'arrow_downward', theme: DEFAULT_THEME };
