// mews-ui/src/Table/TableHeaderCell.jsx
import React, { Component } from 'react';
import { PropTypes as pt } from 'prop-types';
import SortOrder from './SortOrder';
import { TableHeaderCellElement, SortAscendingIcon, SortDescendingIcon } from './style';

class TableHeaderCell extends Component {
    static propTypes = {
        sortable: pt.bool,
        title: pt.string,
        valueKey: pt.string,
        columnAlignment: pt.string,
    };

    handleClick = () => {
        this.props.onSort(this.props.columnIndex);
    };

    render() {
        const { columnAlignment, title, sortable, valueKey, sortedBy, sortOrder } = this.props;

        const cellProps = {
            columnAlignment,
            sortable,
            sortedBy,
            onClick: sortable && this.handleClick,
        };

        let sortIcon;
        if (sortedBy) {
            sortIcon = sortOrder === SortOrder.Ascending ? <SortAscendingIcon /> : <SortDescendingIcon />;
        }

        return (
            <TableHeaderCellElement {...cellProps}>
                {sortIcon}
                {title || valueKey}
            </TableHeaderCellElement>
        );
    }
}

export default TableHeaderCell;
