// mews-js/src/components/finance/paymentGateways/adyenComponent.js
import encrypt from 'adyen-cse-js';
import { type } from '../../../utils';

const DEFAULT_HOLDER_NAME = 'Cardholder';

export default function AdyenComponent() {
    return type.create(AdyenComponent, {
        tokenizeCard(
            {
                number,
                cvv: cvc,
                expiration: {
                    month: expiryMonth,
                    year: expiryYear,
                },
            },
            { publicKey, nowUtc: generationtime }
        ) {
            const nonce = encrypt.createEncryption(publicKey, { enableValidations: false }).encrypt({
                cvc,
                number,
                expiryMonth,
                expiryYear,
                generationtime,
                holderName: DEFAULT_HOLDER_NAME,
            });

            if (!nonce) {
                return Promise.reject(new Error('Adyen gateway client refused to encrypt the card.'));
            }

            return Promise.resolve(nonce);
        },
    });
}
