// mews-js/src/components/finance/paymentGateways/braintreeComponent.js
import client from 'braintree-web/client';
import { type } from '../../../utils';

export default function BraintreeComponent() {
    return type.create(BraintreeComponent, {
        tokenizeCard(
            {
                cvv,
                number,
                expiration: {
                    month: expirationMonth,
                    year: expirationYear,
                },
            },
            { clientToken }
        ) {
            return new Promise((resolve, reject) => {
                client.create({ authorization: clientToken }, (clientErr, clientInstance) => {
                    if (clientErr) {
                        return reject(clientErr);
                    }

                    clientInstance.request({
                        endpoint: 'payment_methods/credit_cards',
                        method: 'post',
                        data: {
                            creditCard: {
                                cvv,
                                expirationMonth,
                                expirationYear,
                                number,
                            },
                        },
                    }, (err, response) => {
                        if (err) {
                            return reject(err);
                        }

                        resolve(response.creditCards[0].nonce);
                    });
                });
            });
        },
    });
}
