// mews-js/src/components/finance/paymentGateways/stripeComponent.js
import { type, loadScript } from '../../../utils';

const STRIPE_LIBRARY_URL = 'https://js.stripe.com/v2';
const DUMMY_CVC = '123';

export default function StripeComponent() {
    return type.create(StripeComponent, {
        tokenizeCard(
            {
                cvv: cvc,
                number,
                expiration: {
                    month: expirationMonth,
                    year: expirationYear,
                },
            },
            { publishableKey }
        ) {
            return Promise.resolve()
                .then(() => global.Stripe || loadScript(STRIPE_LIBRARY_URL))
                .then(() => {
                    global.Stripe.setPublishableKey(publishableKey);

                    return new Promise((resolve, reject) => {
                        global.Stripe.card.createToken({
                            number,
                            cvc: cvc || DUMMY_CVC,
                            exp_month: expirationMonth, // eslint-disable-line camelcase
                            exp_year: expirationYear, // eslint-disable-line camelcase
                        }, (status, response) => {
                            if (response.error) {
                                return reject(response.error);
                            }

                            resolve(response.id);
                        });
                    });
                });
        },
    });
}
