// mews-ui/src/Slider/style.jsx
import React from 'react';
import styled from 'styled-components';
import { COLOR_WHITE } from '../colors';
import { DEFAULT_THEME } from '../variables';
import ProgressLinear from '../ProgressBar/ProgressLinear';
import ProgressMode from '../ProgressBar/ProgressMode';
import { PROGRESS_BAR_COLOR, LinearValueBar } from '../ProgressBar/style';

const SLIDER_HEIGHT = '48px';

export const SliderElement = styled.div`
    cursor: pointer;
    height: ${SLIDER_HEIGHT};
    outline: none;
    position: relative;
    touch-action: none;
    user-select: none;
    width: 100%;
`;

export const SliderTrackElement = styled(ProgressLinear).attrs({
    mode: ProgressMode.Determinate,
})`
    position: absolute;
    left: 0;
    top: 50%;
    height: 2px;

    ${LinearValueBar} {
        transition-duration: 0s;
    }
`;

export const KnobContainerElement = styled.div`
    align-items: center;
    background-color: transparent;
    display: inline-flex;
    height: ${SLIDER_HEIGHT};
    justify-content: center;
    width: 20px;
`;

const KnobSvgElement = styled.svg.attrs({
    width: 21,
    height: 21,
})`
    fill: ${props => props.theme.primaryColor};
    stroke: ${props => props.theme.primaryColor};
    stroke-width: 3.5;
    position: absolute;
    transform: scale(0.571);
    transition: transform 0.1s ease-out, fill 0.1s ease-out, stroke 0.1s ease-out;

    ${props => props.active && `
        transform: scale3d(1, 1, 1);
    `}

    ${props => props.disabled && `
        stroke: ${COLOR_WHITE};
        fill: ${PROGRESS_BAR_COLOR};
    `}
`;
KnobSvgElement.defaultProps = { theme: DEFAULT_THEME };

const KnobCircleElement = styled.circle.attrs({
    cx: 10.5,
    cy: 10.5,
    r: 7.875,
})``;

export const KnobElement = props => (
    <KnobSvgElement {...props}>
        <KnobCircleElement />
    </KnobSvgElement>
);

export const KnobFocusElement = styled.div`
    background-color: ${props => props.theme.primaryColor};
    border-radius: 50%;
    height: 21px;
    opacity: 0;
    width: 21px;

    transition: transform 266.67ms ease-out, opacity 266.67ms ease-out, background-color 266.67ms ease-out;
    
    ${props => props.focused && `
        transform: scale3d(1.55, 1.55, 1.55);
        opacity: .25;
    `}
`;
KnobFocusElement.defaultProps = { theme: DEFAULT_THEME };
