// mews-ui/lib/Table/style.js
var _templateObject = _taggedTemplateLiteral(['\n    border-spacing: 0;\n    width: 100%;\n'], ['\n    border-spacing: 0;\n    width: 100%;\n']),
    _templateObject2 = _taggedTemplateLiteral([''], ['']),
    _templateObject3 = _taggedTemplateLiteral(['\n    ', '\n'], ['\n    ', '\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n    font-size: 12px;\n    font-weight: 500;\n    color: ', ';\n'], ['\n    font-size: 12px;\n    font-weight: 500;\n    color: ', ';\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n    font-size: 13px;\n    color: ', ';\n'], ['\n    font-size: 13px;\n    color: ', ';\n']),
    _templateObject6 = _taggedTemplateLiteral(['\n    padding: 24px 0;\n'], ['\n    padding: 24px 0;\n']),
    _templateObject7 = _taggedTemplateLiteral(['\n    border-bottom: 1px solid ', ';\n    box-sizing: border-box;\n    height: 56px;\n    padding: 0;\n    text-align: ', ';\n\n    ', '\n'], ['\n    border-bottom: 1px solid ', ';\n    box-sizing: border-box;\n    height: 56px;\n    padding: 0;\n    text-align: ', ';\n\n    ', '\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n    border-bottom: 1px solid ', ';\n    box-sizing: border-box;\n    height: 48px;\n    padding: 0;\n    text-align: ', ';\n'], ['\n    border-bottom: 1px solid ', ';\n    box-sizing: border-box;\n    height: 48px;\n    padding: 0;\n    text-align: ', ';\n']),
    _templateObject9 = _taggedTemplateLiteral(['\n    color: ', '\n    font-size: 16px;\n    vertical-align: text-bottom;\n'], ['\n    color: ', '\n    font-size: 16px;\n    vertical-align: text-bottom;\n']),
    _templateObject10 = _taggedTemplateLiteral(['', ''], ['', '']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import styled, { css } from 'styled-components';
import { DEFAULT_THEME } from '../variables';
import FontIcon from '../FontIcon/FontIcon';

export var TableElement = styled.table(_templateObject);

export var ColumnDefinitionsElement = styled.colgroup(_templateObject2);

export var ColumnElement = styled.col(_templateObject3, function (props) {
    return props.columnWidth && '\n        width: ' + props.columnWidth + ';\n    ';
});

export var TableHeadElement = styled.thead(_templateObject4, function (props) {
    return props.theme.primaryTextColor;
});
TableHeadElement.defaultProps = { theme: DEFAULT_THEME };

export var TableBodyElement = styled.tbody(_templateObject5, function (props) {
    return props.theme.primaryTextColor;
});
TableBodyElement.defaultProps = { theme: DEFAULT_THEME };

export var TableHeaderRowElement = styled.tr(_templateObject6);

export var TableHeaderCellElement = styled.th(_templateObject7, function (props) {
    return props.theme.dividerColor;
}, function (props) {
    return props.columnAlignment || 'left';
}, function (props) {
    return props.sortable && '\n        cursor: pointer;\n    ';
});
TableHeaderCellElement.defaultProps = { theme: DEFAULT_THEME };

export var TableRowElement = styled.tr(_templateObject2);

export var TableCellElement = styled.td(_templateObject8, function (props) {
    return props.theme.dividerColor;
}, function (props) {
    return props.columnAlignment || 'left';
});
TableCellElement.defaultProps = { theme: DEFAULT_THEME };

var iconStyle = css(_templateObject9, function (props) {
    return props.theme.primaryTextColor;
});

export var SortAscendingIcon = styled(FontIcon)(_templateObject10, iconStyle);
SortAscendingIcon.defaultProps = { value: 'arrow_upward', theme: DEFAULT_THEME };

export var SortDescendingIcon = styled(FontIcon)(_templateObject10, iconStyle);
SortDescendingIcon.defaultProps = { value: 'arrow_downward', theme: DEFAULT_THEME };
