// mews-ui/src/colors.js
export const PALETTE_RED_50 = 'rgb(255, 235, 238)';
export const PALETTE_RED_100 = 'rgb(255, 205, 210)';
export const PALETTE_RED_200 = 'rgb(239, 154, 154)';
export const PALETTE_RED_300 = 'rgb(229, 115, 115)';
export const PALETTE_RED_400 = 'rgb(239, 83, 80)';
export const PALETTE_RED_500 = 'rgb(244, 67, 54)';
export const PALETTE_RED_600 = 'rgb(229, 57, 53)';
export const PALETTE_RED_700 = 'rgb(211, 47, 47)';
export const PALETTE_RED_800 = 'rgb(198, 40, 40)';
export const PALETTE_RED_900 = 'rgb(183, 28, 28)';
export const PALETTE_RED_A100 = 'rgb(255, 138, 128)';
export const PALETTE_RED_A200 = 'rgb(255, 82, 82)';
export const PALETTE_RED_A400 = 'rgb(255, 23, 68)';
export const PALETTE_RED_A700 = 'rgb(213, 0, 0)';

export const PALETTE_PINK_50 = 'rgb(252, 228, 236)';
export const PALETTE_PINK_100 = 'rgb(248, 187, 208)';
export const PALETTE_PINK_200 = 'rgb(244, 143, 177)';
export const PALETTE_PINK_300 = 'rgb(240, 98, 146)';
export const PALETTE_PINK_400 = 'rgb(236, 64, 122)';
export const PALETTE_PINK_500 = 'rgb(233, 30, 99)';
export const PALETTE_PINK_600 = 'rgb(216, 27, 96)';
export const PALETTE_PINK_700 = 'rgb(194, 24, 91)';
export const PALETTE_PINK_800 = 'rgb(173, 20, 87)';
export const PALETTE_PINK_900 = 'rgb(136, 14, 79)';
export const PALETTE_PINK_A100 = 'rgb(255, 128, 171)';
export const PALETTE_PINK_A200 = 'rgb(255, 64, 129)';
export const PALETTE_PINK_A400 = 'rgb(245, 0, 87)';
export const PALETTE_PINK_A700 = 'rgb(197, 17, 98)';

export const PALETTE_PURPLE_50 = 'rgb(243, 229, 245)';
export const PALETTE_PURPLE_100 = 'rgb(225, 190, 231)';
export const PALETTE_PURPLE_200 = 'rgb(206, 147, 216)';
export const PALETTE_PURPLE_300 = 'rgb(186, 104, 200)';
export const PALETTE_PURPLE_400 = 'rgb(171, 71, 188)';
export const PALETTE_PURPLE_500 = 'rgb(156, 39, 176)';
export const PALETTE_PURPLE_600 = 'rgb(142, 36, 170)';
export const PALETTE_PURPLE_700 = 'rgb(123, 31, 162)';
export const PALETTE_PURPLE_800 = 'rgb(106, 27, 154)';
export const PALETTE_PURPLE_900 = 'rgb(74, 20, 140)';
export const PALETTE_PURPLE_A100 = 'rgb(234, 128, 252)';
export const PALETTE_PURPLE_A200 = 'rgb(224, 64, 251)';
export const PALETTE_PURPLE_A400 = 'rgb(213, 0, 249)';
export const PALETTE_PURPLE_A700 = 'rgb(170, 0, 255)';

export const PALETTE_DEEP_PURPLE_50 = 'rgb(237, 231, 246)';
export const PALETTE_DEEP_PURPLE_100 = 'rgb(209, 196, 233)';
export const PALETTE_DEEP_PURPLE_200 = 'rgb(179, 157, 219)';
export const PALETTE_DEEP_PURPLE_300 = 'rgb(149, 117, 205)';
export const PALETTE_DEEP_PURPLE_400 = 'rgb(126, 87, 194)';
export const PALETTE_DEEP_PURPLE_500 = 'rgb(103, 58, 183)';
export const PALETTE_DEEP_PURPLE_600 = 'rgb(94, 53, 177)';
export const PALETTE_DEEP_PURPLE_700 = 'rgb(81, 45, 168)';
export const PALETTE_DEEP_PURPLE_800 = 'rgb(69, 39, 160)';
export const PALETTE_DEEP_PURPLE_900 = 'rgb(49, 27, 146)';
export const PALETTE_DEEP_PURPLE_A100 = 'rgb(179, 136, 255)';
export const PALETTE_DEEP_PURPLE_A200 = 'rgb(124, 77, 255)';
export const PALETTE_DEEP_PURPLE_A400 = 'rgb(101, 31, 255)';
export const PALETTE_DEEP_PURPLE_A700 = 'rgb(98, 0, 234)';

export const PALETTE_INDIGO_50 = 'rgb(232, 234, 246)';
export const PALETTE_INDIGO_100 = 'rgb(197, 202, 233)';
export const PALETTE_INDIGO_200 = 'rgb(159, 168, 218)';
export const PALETTE_INDIGO_300 = 'rgb(121, 134, 203)';
export const PALETTE_INDIGO_400 = 'rgb(92, 107, 192)';
export const PALETTE_INDIGO_500 = 'rgb(63, 81, 181)';
export const PALETTE_INDIGO_600 = 'rgb(57, 73, 171)';
export const PALETTE_INDIGO_700 = 'rgb(48, 63, 159)';
export const PALETTE_INDIGO_800 = 'rgb(40, 53, 147)';
export const PALETTE_INDIGO_900 = 'rgb(26, 35, 126)';
export const PALETTE_INDIGO_A100 = 'rgb(140, 158, 255)';
export const PALETTE_INDIGO_A200 = 'rgb(83, 109, 254)';
export const PALETTE_INDIGO_A400 = 'rgb(61, 90, 254)';
export const PALETTE_INDIGO_A700 = 'rgb(48, 79, 254)';

export const PALETTE_BLUE_50 = 'rgb(227, 242, 253)';
export const PALETTE_BLUE_100 = 'rgb(187, 222, 251)';
export const PALETTE_BLUE_200 = 'rgb(144, 202, 249)';
export const PALETTE_BLUE_300 = 'rgb(100, 181, 246)';
export const PALETTE_BLUE_400 = 'rgb(66, 165, 245)';
export const PALETTE_BLUE_500 = 'rgb(33, 150, 243)';
export const PALETTE_BLUE_600 = 'rgb(30, 136, 229)';
export const PALETTE_BLUE_700 = 'rgb(25, 118, 210)';
export const PALETTE_BLUE_800 = 'rgb(21, 101, 192)';
export const PALETTE_BLUE_900 = 'rgb(13, 71, 161)';
export const PALETTE_BLUE_A100 = 'rgb(130, 177, 255)';
export const PALETTE_BLUE_A200 = 'rgb(68, 138, 255)';
export const PALETTE_BLUE_A400 = 'rgb(41, 121, 255)';
export const PALETTE_BLUE_A700 = 'rgb(41, 98, 255)';

export const PALETTE_LIGHT_BLUE_50 = 'rgb(225, 245, 254)';
export const PALETTE_LIGHT_BLUE_100 = 'rgb(179, 229, 252)';
export const PALETTE_LIGHT_BLUE_200 = 'rgb(129, 212, 250)';
export const PALETTE_LIGHT_BLUE_300 = 'rgb(79, 195, 247)';
export const PALETTE_LIGHT_BLUE_400 = 'rgb(41, 182, 246)';
export const PALETTE_LIGHT_BLUE_500 = 'rgb(3, 169, 244)';
export const PALETTE_LIGHT_BLUE_600 = 'rgb(3, 155, 229)';
export const PALETTE_LIGHT_BLUE_700 = 'rgb(2, 136, 209)';
export const PALETTE_LIGHT_BLUE_800 = 'rgb(2, 119, 189)';
export const PALETTE_LIGHT_BLUE_900 = 'rgb(1, 87, 155)';
export const PALETTE_LIGHT_BLUE_A100 = 'rgb(128, 216, 255)';
export const PALETTE_LIGHT_BLUE_A200 = 'rgb(64, 196, 255)';
export const PALETTE_LIGHT_BLUE_A400 = 'rgb(0, 176, 255)';
export const PALETTE_LIGHT_BLUE_A700 = 'rgb(0, 145, 234)';

export const PALETTE_CYAN_50 = 'rgb(224, 247, 250)';
export const PALETTE_CYAN_100 = 'rgb(178, 235, 242)';
export const PALETTE_CYAN_200 = 'rgb(128, 222, 234)';
export const PALETTE_CYAN_300 = 'rgb(77, 208, 225)';
export const PALETTE_CYAN_400 = 'rgb(38, 198, 218)';
export const PALETTE_CYAN_500 = 'rgb(0, 188, 212)';
export const PALETTE_CYAN_600 = 'rgb(0, 172, 193)';
export const PALETTE_CYAN_700 = 'rgb(0, 151, 167)';
export const PALETTE_CYAN_800 = 'rgb(0, 131, 143)';
export const PALETTE_CYAN_900 = 'rgb(0, 96, 100)';
export const PALETTE_CYAN_A100 = 'rgb(132, 255, 255)';
export const PALETTE_CYAN_A200 = 'rgb(24, 255, 255)';
export const PALETTE_CYAN_A400 = 'rgb(0, 229, 255)';
export const PALETTE_CYAN_A700 = 'rgb(0, 184, 212)';

export const PALETTE_TEAL_50 = 'rgb(224, 242, 241)';
export const PALETTE_TEAL_100 = 'rgb(178, 223, 219)';
export const PALETTE_TEAL_200 = 'rgb(128, 203, 196)';
export const PALETTE_TEAL_300 = 'rgb(77, 182, 172)';
export const PALETTE_TEAL_400 = 'rgb(38, 166, 154)';
export const PALETTE_TEAL_500 = 'rgb(0, 150, 136)';
export const PALETTE_TEAL_600 = 'rgb(0, 137, 123)';
export const PALETTE_TEAL_700 = 'rgb(0, 121, 107)';
export const PALETTE_TEAL_800 = 'rgb(0, 105, 92)';
export const PALETTE_TEAL_900 = 'rgb(0, 77, 64)';
export const PALETTE_TEAL_A100 = 'rgb(167, 255, 235)';
export const PALETTE_TEAL_A200 = 'rgb(100, 255, 218)';
export const PALETTE_TEAL_A400 = 'rgb(29, 233, 182)';
export const PALETTE_TEAL_A700 = 'rgb(0, 191, 165)';

export const PALETTE_GREEN_50 = 'rgb(232, 245, 233)';
export const PALETTE_GREEN_100 = 'rgb(200, 230, 201)';
export const PALETTE_GREEN_200 = 'rgb(165, 214, 167)';
export const PALETTE_GREEN_300 = 'rgb(129, 199, 132)';
export const PALETTE_GREEN_400 = 'rgb(102, 187, 106)';
export const PALETTE_GREEN_500 = 'rgb(76, 175, 80)';
export const PALETTE_GREEN_600 = 'rgb(67, 160, 71)';
export const PALETTE_GREEN_700 = 'rgb(56, 142, 60)';
export const PALETTE_GREEN_800 = 'rgb(46, 125, 50)';
export const PALETTE_GREEN_900 = 'rgb(27, 94, 32)';
export const PALETTE_GREEN_A100 = 'rgb(185, 246, 202)';
export const PALETTE_GREEN_A200 = 'rgb(105, 240, 174)';
export const PALETTE_GREEN_A400 = 'rgb(0, 230, 118)';
export const PALETTE_GREEN_A700 = 'rgb(0, 200, 83)';

export const PALETTE_LIGHT_GREEN_50 = 'rgb(241, 248, 233)';
export const PALETTE_LIGHT_GREEN_100 = 'rgb(220, 237, 200)';
export const PALETTE_LIGHT_GREEN_200 = 'rgb(197, 225, 165)';
export const PALETTE_LIGHT_GREEN_300 = 'rgb(174, 213, 129)';
export const PALETTE_LIGHT_GREEN_400 = 'rgb(156, 204, 101)';
export const PALETTE_LIGHT_GREEN_500 = 'rgb(139, 195, 74)';
export const PALETTE_LIGHT_GREEN_600 = 'rgb(124, 179, 66)';
export const PALETTE_LIGHT_GREEN_700 = 'rgb(104, 159, 56)';
export const PALETTE_LIGHT_GREEN_800 = 'rgb(85, 139, 47)';
export const PALETTE_LIGHT_GREEN_900 = 'rgb(51, 105, 30)';
export const PALETTE_LIGHT_GREEN_A100 = 'rgb(204, 255, 144)';
export const PALETTE_LIGHT_GREEN_A200 = 'rgb(178, 255, 89)';
export const PALETTE_LIGHT_GREEN_A400 = 'rgb(118, 255, 3)';
export const PALETTE_LIGHT_GREEN_A700 = 'rgb(100, 221, 23)';

export const PALETTE_LIME_50 = 'rgb(249, 251, 231)';
export const PALETTE_LIME_100 = 'rgb(240, 244, 195)';
export const PALETTE_LIME_200 = 'rgb(230, 238, 156)';
export const PALETTE_LIME_300 = 'rgb(220, 231, 117)';
export const PALETTE_LIME_400 = 'rgb(212, 225, 87)';
export const PALETTE_LIME_500 = 'rgb(205, 220, 57)';
export const PALETTE_LIME_600 = 'rgb(192, 202, 51)';
export const PALETTE_LIME_700 = 'rgb(175, 180, 43)';
export const PALETTE_LIME_800 = 'rgb(158, 157, 36)';
export const PALETTE_LIME_900 = 'rgb(130, 119, 23)';
export const PALETTE_LIME_A100 = 'rgb(244, 255, 129)';
export const PALETTE_LIME_A200 = 'rgb(238, 255, 65)';
export const PALETTE_LIME_A400 = 'rgb(198, 255, 0)';
export const PALETTE_LIME_A700 = 'rgb(174, 234, 0)';

export const PALETTE_YELLOW_50 = 'rgb(255, 253, 231)';
export const PALETTE_YELLOW_100 = 'rgb(255, 249, 196)';
export const PALETTE_YELLOW_200 = 'rgb(255, 245, 157)';
export const PALETTE_YELLOW_300 = 'rgb(255, 241, 118)';
export const PALETTE_YELLOW_400 = 'rgb(255, 238, 88)';
export const PALETTE_YELLOW_500 = 'rgb(255, 235, 59)';
export const PALETTE_YELLOW_600 = 'rgb(253, 216, 53)';
export const PALETTE_YELLOW_700 = 'rgb(251, 192, 45)';
export const PALETTE_YELLOW_800 = 'rgb(249, 168, 37)';
export const PALETTE_YELLOW_900 = 'rgb(245, 127, 23)';
export const PALETTE_YELLOW_A100 = 'rgb(255, 255, 141)';
export const PALETTE_YELLOW_A200 = 'rgb(255, 255, 0)';
export const PALETTE_YELLOW_A400 = 'rgb(255, 234, 0)';
export const PALETTE_YELLOW_A700 = 'rgb(255, 214, 0)';

export const PALETTE_AMBER_50 = 'rgb(255, 248, 225)';
export const PALETTE_AMBER_100 = 'rgb(255, 236, 179)';
export const PALETTE_AMBER_200 = 'rgb(255, 224, 130)';
export const PALETTE_AMBER_300 = 'rgb(255, 213, 79)';
export const PALETTE_AMBER_400 = 'rgb(255, 202, 40)';
export const PALETTE_AMBER_500 = 'rgb(255, 193, 7)';
export const PALETTE_AMBER_600 = 'rgb(255, 179, 0)';
export const PALETTE_AMBER_700 = 'rgb(255, 160, 0)';
export const PALETTE_AMBER_800 = 'rgb(255, 143, 0)';
export const PALETTE_AMBER_900 = 'rgb(255, 111, 0)';
export const PALETTE_AMBER_A100 = 'rgb(255, 229, 127)';
export const PALETTE_AMBER_A200 = 'rgb(255, 215, 64)';
export const PALETTE_AMBER_A400 = 'rgb(255, 196, 0)';
export const PALETTE_AMBER_A700 = 'rgb(255, 171, 0)';

export const PALETTE_ORANGE_50 = 'rgb(255, 243, 224)';
export const PALETTE_ORANGE_100 = 'rgb(255, 224, 178)';
export const PALETTE_ORANGE_200 = 'rgb(255, 204, 128)';
export const PALETTE_ORANGE_300 = 'rgb(255, 183, 77)';
export const PALETTE_ORANGE_400 = 'rgb(255, 167, 38)';
export const PALETTE_ORANGE_500 = 'rgb(255, 152, 0)';
export const PALETTE_ORANGE_600 = 'rgb(251, 140, 0)';
export const PALETTE_ORANGE_700 = 'rgb(245, 124, 0)';
export const PALETTE_ORANGE_800 = 'rgb(239, 108, 0)';
export const PALETTE_ORANGE_900 = 'rgb(230, 81, 0)';
export const PALETTE_ORANGE_A100 = 'rgb(255, 209, 128)';
export const PALETTE_ORANGE_A200 = 'rgb(255, 171, 64)';
export const PALETTE_ORANGE_A400 = 'rgb(255, 145, 0)';
export const PALETTE_ORANGE_A700 = 'rgb(255, 109, 0)';

export const PALETTE_DEEP_ORANGE_50 = 'rgb(251, 233, 231)';
export const PALETTE_DEEP_ORANGE_100 = 'rgb(255, 204, 188)';
export const PALETTE_DEEP_ORANGE_200 = 'rgb(255, 171, 145)';
export const PALETTE_DEEP_ORANGE_300 = 'rgb(255, 138, 101)';
export const PALETTE_DEEP_ORANGE_400 = 'rgb(255, 112, 67)';
export const PALETTE_DEEP_ORANGE_500 = 'rgb(255, 87, 34)';
export const PALETTE_DEEP_ORANGE_600 = 'rgb(244, 81, 30)';
export const PALETTE_DEEP_ORANGE_700 = 'rgb(230, 74, 25)';
export const PALETTE_DEEP_ORANGE_800 = 'rgb(216, 67, 21)';
export const PALETTE_DEEP_ORANGE_900 = 'rgb(191, 54, 12)';
export const PALETTE_DEEP_ORANGE_A100 = 'rgb(255, 158, 128)';
export const PALETTE_DEEP_ORANGE_A200 = 'rgb(255, 110, 64)';
export const PALETTE_DEEP_ORANGE_A400 = 'rgb(255, 61, 0)';
export const PALETTE_DEEP_ORANGE_A700 = 'rgb(221, 44, 0)';

export const PALETTE_BROWN_50 = 'rgb(239, 235, 233)';
export const PALETTE_BROWN_100 = 'rgb(215, 204, 200)';
export const PALETTE_BROWN_200 = 'rgb(188, 170, 164)';
export const PALETTE_BROWN_300 = 'rgb(161, 136, 127)';
export const PALETTE_BROWN_400 = 'rgb(141, 110, 99)';
export const PALETTE_BROWN_500 = 'rgb(121, 85, 72)';
export const PALETTE_BROWN_600 = 'rgb(109, 76, 65)';
export const PALETTE_BROWN_700 = 'rgb(93, 64, 55)';
export const PALETTE_BROWN_800 = 'rgb(78, 52, 46)';
export const PALETTE_BROWN_900 = 'rgb(62, 39, 35)';

export const PALETTE_GREY_50 = 'rgb(250, 250, 250)';
export const PALETTE_GREY_100 = 'rgb(245, 245, 245)';
export const PALETTE_GREY_200 = 'rgb(238, 238, 238)';
export const PALETTE_GREY_300 = 'rgb(224, 224, 224)';
export const PALETTE_GREY_400 = 'rgb(189, 189, 189)';
export const PALETTE_GREY_500 = 'rgb(158, 158, 158)';
export const PALETTE_GREY_600 = 'rgb(117, 117, 117)';
export const PALETTE_GREY_700 = 'rgb(97, 97, 97)';
export const PALETTE_GREY_800 = 'rgb(66, 66, 66)';
export const PALETTE_GREY_900 = 'rgb(33, 33, 33)';

export const PALETTE_BLUE_GREY_50 = 'rgb(236, 239, 241)';
export const PALETTE_BLUE_GREY_100 = 'rgb(207, 216, 220)';
export const PALETTE_BLUE_GREY_200 = 'rgb(176, 190, 197)';
export const PALETTE_BLUE_GREY_300 = 'rgb(144, 164, 174)';
export const PALETTE_BLUE_GREY_400 = 'rgb(120, 144, 156)';
export const PALETTE_BLUE_GREY_500 = 'rgb(96, 125, 139)';
export const PALETTE_BLUE_GREY_600 = 'rgb(84, 110, 122)';
export const PALETTE_BLUE_GREY_700 = 'rgb(69, 90, 100)';
export const PALETTE_BLUE_GREY_800 = 'rgb(55, 71, 79)';
export const PALETTE_BLUE_GREY_900 = 'rgb(38, 50, 56)';

export const COLOR_BLACK = 'rgb(0, 0, 0)';
export const COLOR_WHITE = 'rgb(255, 255, 255)';
export const COLOR_DARK_CONTRAST = COLOR_WHITE;
export const COLOR_LIGHT_CONTRAST = COLOR_BLACK;
export const COLOR_RED = 'rgb(222, 50, 38)';
