// mews-ui/src/variables.js
import {
    COLOR_DARK_CONTRAST,
    COLOR_BLACK,
    PALETTE_BLUE_500,
    PALETTE_GREY_200,
    PALETTE_GREY_600,
    PALETTE_GREY_900,
} from './colors';
import { alpha } from './utils/color';

export const COLOR_DIVIDER = PALETTE_GREY_200;
export const COLOR_TEXT = PALETTE_GREY_900;
export const COLOR_TEXT_SECONDARY = PALETTE_GREY_600;

export const COLOR_PRIMARY = PALETTE_BLUE_500;
export const COLOR_PRIMARY_CONTRAST = COLOR_DARK_CONTRAST;

export const PREFERRED_FONT = "'Roboto', 'Helvetica', 'Arial', sans-serif";
export const FONT_SIZE_TINY = '12px';
export const FONT_SIZE_SMALL = '14px';
export const FONT_SIZE_NORMAL = '16px';
export const FONT_SIZE_LARGE = '25px';
export const FONT_SIZE = FONT_SIZE_NORMAL;
export const FONT_WEIGHT_THIN = 300;

export const ANIMATION_DURATION = '0.35s';
export const ANIMATION_CURVE_FAST_OUT_LINEAR_IN = 'cubic-bezier(.4, 0, 1, 1)';
export const ANIMATION_CURVE_FAST_OUT_SLOW_IN = 'cubic-bezier(.4, 0, .2, 1)';
export const ANIMATION_CURVE_LINEAR_OUT_SLOW_IN = 'cubic-bezier(0, 0, .2, 1)';
export const ANIMATION_CURVE_DEFAULT = ANIMATION_CURVE_FAST_OUT_SLOW_IN;

export const SHADOW_2P = `
    0 2px 2px 0 rgba(0, 0, 0, 0.14),
    0 3px 1px -2px rgba(0, 0, 0, 0.2),
    0 1px 5px 0 rgba(0, 0, 0, 0.12);
`;

export const SHADOW_3P = `
    0 3px 4px 0 rgba(0, 0, 0, 0.14),
    0 3px 3px -2px rgba(0, 0, 0, 0.2),
    0 1px 8px 0 rgba(0, 0, 0, 0.12);
`;

export const Z_DEPTH_SHADOW_1 = '0 1px 6px rgba(0, 0, 0, 0.12), 0 1px 4px rgba(0, 0, 0, 0.24)';
export const Z_DEPTH_SHADOW_2 = '0 3px 10px rgba(0, 0, 0, 0.16), 0 3px 10px rgba(0, 0, 0, 0.23)';
export const Z_DEPTH_SHADOW_3 = '0 10px 30px rgba(0, 0, 0, 0.19), 0 6px 10px rgba(0, 0, 0, 0.23)';
export const Z_DEPTH_SHADOW_4 = '0 14px 45px rgba(0, 0, 0, 0.25), 0 10px 18px rgba(0, 0, 0, 0.22)';
export const Z_DEPTH_SHADOW_5 = '0 19px 60px rgba(0, 0, 0, 0.3), 0 15px 20px rgba(0, 0, 0, 0.22)';

export const Z_INDEX_HIGHEST = 300;
export const Z_INDEX_HIGHER = 200;
export const Z_INDEX_HIGH = 100;
export const Z_INDEX_NORMAL = 1;
export const Z_INDEX_LOW = -100;
export const Z_INDEX_LOWER = -200;

const disabledColor = alpha(COLOR_BLACK, 0.38);
const disabledSwitchColor = '#fafafa';

export const DEFAULT_THEME = {
    primaryColor: COLOR_PRIMARY,
    primaryContrastColor: COLOR_PRIMARY_CONTRAST,

    primaryTextColor: alpha(COLOR_BLACK, 0.87),
    secondaryTextColor: alpha(COLOR_BLACK, 0.54),
    disabledColor,
    disabledSwitchColor,
    hintTextColor: disabledColor,
    iconColor: disabledColor,
    dividerColor: alpha(COLOR_BLACK, 0.12),
};
