// mews-ui/src/List/style.jsx
import styled, { css } from 'styled-components';
import { PALETTE_GREY_200 } from '../colors';
import { COLOR_TEXT_SECONDARY, FONT_SIZE_SMALL, FONT_SIZE_NORMAL, DEFAULT_THEME } from '../variables';
import { InputElement } from '../Input/style';

const LIST_VERTICAL_PADDING = props => (props.dense ? '4px' : '8px');
const LIST_HORIZONTAL_PADDING = '16px';
const LIST_ITEM_HOVER_COLOR = PALETTE_GREY_200;
const LIST_ITEM_ACTIVE_BACKGROUND = PALETTE_GREY_200;

export const ListElement = styled.div`
    outline: none;
    padding: ${LIST_VERTICAL_PADDING} 0;
    position: relative;
    text-align: left;
`;

export const ListItemsWrapperElement = styled.div`
    outline: none;
`;

export const ListItemElement = styled.div`
    align-items: center;
    box-sizing: border-box;
    display: flex;
    outline: none;
    padding: ${LIST_VERTICAL_PADDING} ${LIST_HORIZONTAL_PADDING};
    position: relative;
    height: ${props => (props.dense ? '40px' : '48px')};

    ${props => props.singleLineWithAvatar && css`
        height: ${props.dense ? '48px' : '56px'};
    `}

    ${props => props.twoLine && css`
        height: ${props.dense ? '60px' : '72px'};
    `}

    ${props => props.threeLine && css`
        height: ${props.dense ? '76px' : '88px'};
    `}

    ${props => props.selected && css`
        color: ${props.theme.primaryColor};
    `}

    ${props => props.hidden && css`
        display: none;
    `}

    &:focus {
        background-color: ${LIST_ITEM_ACTIVE_BACKGROUND};
    }

    &:hover {
        background-color: ${LIST_ITEM_HOVER_COLOR};
        cursor: pointer;
    }
`;
ListItemElement.defaultProps = {
    theme: DEFAULT_THEME,
};

export const ListItemWrapper = styled.span`
    outline: none;
`;

export const ListItemLeftElement = styled.span`
    align-items: center;
    display: flex;
`;

export const ListItemRightElement = styled.span`
    align-items: center;
    display: flex;
    margin-left: auto;
    padding-left: ${LIST_HORIZONTAL_PADDING};
`;

export const ListItemIconElement = styled.span`
    color: ${COLOR_TEXT_SECONDARY};
    
    ${props => props.left && `margin-right: 32px;`}
    ${props => props.right && `margin-left: 16px;`}
`;

export const ListItemAvatarElement = styled.span`
    margin-right: 16px;
`;

export const ListItemContentElement = styled.span`
    display: flex;
    flex-flow: column;
    overflow-x: hidden;
`;

export const ListItemPrimaryTextElement = styled.span`
    font-size: ${props => (props.dense ? FONT_SIZE_SMALL : FONT_SIZE_NORMAL)};
    overflow-x: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
`;

export const ListItemSecondaryTextElement = styled.span`
    color: ${COLOR_TEXT_SECONDARY};
    font-size: ${FONT_SIZE_SMALL};
    overflow-x: hidden;
    text-overflow: ellipsis;
    word-wrap: nowrap;
`;

export const FilterElement = styled.div`
    position: relative;
    border-bottom: 1px solid ${props => props.theme.dividerColor};
    padding: 8px 16px;
`;
FilterElement.defaultProps = { theme: DEFAULT_THEME };
export const FilterInputElement = InputElement;
export const FilterPlaceholderElement = styled.div`
    position: absolute;
    top: 8px;
    color: ${props => props.theme.hintTextColor};
    pointer-events: none;
`;
FilterPlaceholderElement.defaultProps = { theme: DEFAULT_THEME };
